{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data loading and visualization\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from bioinfokit import analys, visuz\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "from sklearn import preprocessing\n",
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Kuehn et al. (2024)\n",
    "\n",
    "RNASeq analysis for EBV-immortalized B-cells from individuals with an AIOLOS E82K mutation (n=4) and healthy controls (n=3)\n",
    "- *kuehn_data/GSE242931_genematrix.txt*\n",
    "\n",
    "We are interested in EBV-immortalized B cells from healthy controls only. \n",
    "\n",
    "\"We used reference **genome hg19**, and all files necessary for analysis were taken from the platform/resource **UCSC Genome Browser** (University of California, Santa Cruz, USA).\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ensemble</th>\n",
       "      <th>HC1_1</th>\n",
       "      <th>HC2_1</th>\n",
       "      <th>HC3_1</th>\n",
       "      <th>B.I.1_1</th>\n",
       "      <th>B.II.2_1</th>\n",
       "      <th>C.II.1_1</th>\n",
       "      <th>C.III.1_1</th>\n",
       "      <th>HC1_2</th>\n",
       "      <th>HC2_2</th>\n",
       "      <th>HC3_2</th>\n",
       "      <th>B.I.1_2</th>\n",
       "      <th>B.II.2_2</th>\n",
       "      <th>C.II.1_2</th>\n",
       "      <th>C.III.1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>17</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000225630</td>\n",
       "      <td>1134</td>\n",
       "      <td>1195</td>\n",
       "      <td>657</td>\n",
       "      <td>1223</td>\n",
       "      <td>671</td>\n",
       "      <td>2475</td>\n",
       "      <td>1992</td>\n",
       "      <td>1199</td>\n",
       "      <td>790</td>\n",
       "      <td>750</td>\n",
       "      <td>974</td>\n",
       "      <td>536</td>\n",
       "      <td>2189</td>\n",
       "      <td>794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>88</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>98</td>\n",
       "      <td>57</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>71</td>\n",
       "      <td>44</td>\n",
       "      <td>25</td>\n",
       "      <td>66</td>\n",
       "      <td>44</td>\n",
       "      <td>42</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000229344</td>\n",
       "      <td>25</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>25</td>\n",
       "      <td>29</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000248527</td>\n",
       "      <td>6605</td>\n",
       "      <td>6262</td>\n",
       "      <td>4856</td>\n",
       "      <td>6234</td>\n",
       "      <td>3118</td>\n",
       "      <td>5544</td>\n",
       "      <td>4551</td>\n",
       "      <td>8658</td>\n",
       "      <td>3438</td>\n",
       "      <td>4987</td>\n",
       "      <td>5283</td>\n",
       "      <td>3060</td>\n",
       "      <td>4876</td>\n",
       "      <td>1557</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_ensemble  HC1_1   HC2_1   HC3_1   B.I.1_1   B.II.2_1   C.II.1_1  \\\n",
       "0  ENSG00000279457     17       6       8         5         12          4   \n",
       "1  ENSG00000225630   1134    1195     657      1223        671       2475   \n",
       "2  ENSG00000237973     88      79      37        98         57         40   \n",
       "3  ENSG00000229344     25      20      23        20          2         21   \n",
       "4  ENSG00000248527   6605    6262    4856      6234       3118       5544   \n",
       "\n",
       "    C.III.1_1   HC1_2   HC2_2   HC3_2   B.I.1_2   B.II.2_2   C.II.1_2  \\\n",
       "0           6      12      10       6        17          3         12   \n",
       "1        1992    1199     790     750       974        536       2189   \n",
       "2          32      71      44      25        66         44         42   \n",
       "3           9      35      21      25        29         18         15   \n",
       "4        4551    8658    3438    4987      5283       3060       4876   \n",
       "\n",
       "    C.III.1_2  \n",
       "0           3  \n",
       "1         794  \n",
       "2          28  \n",
       "3           5  \n",
       "4        1557  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load kuehn et al. RNAseq data\n",
    "kuehn_df = pd.read_csv('../kuehn_data/GSE242931_genematrix.txt', delimiter='\\t', header=0)\n",
    "kuehn_df.rename(columns = {list(kuehn_df)[0]: 'gene_ensemble'}, inplace = True)\n",
    "kuehn_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_ensemble', 'HC1_1', 'HC2_1', 'HC3_1', 'B.I.1_1', 'B.II.2_1',\n",
       "       'C.II.1_1', 'C.III.1_1', 'HC1_2', 'HC2_2', 'HC3_2', 'B.I.1_2',\n",
       "       'B.II.2_2', 'C.II.1_2', 'C.III.1_2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove weird spaces in column names\n",
    "kuehn_df.rename(columns=lambda x: x.strip(), inplace=True)\n",
    "kuehn_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 15870 entries, 0 to 15869\n",
      "Data columns (total 15 columns):\n",
      " #   Column         Non-Null Count  Dtype \n",
      "---  ------         --------------  ----- \n",
      " 0   gene_ensemble  15870 non-null  object\n",
      " 1   HC1_1          15870 non-null  int64 \n",
      " 2   HC2_1          15870 non-null  int64 \n",
      " 3   HC3_1          15870 non-null  int64 \n",
      " 4   B.I.1_1        15870 non-null  int64 \n",
      " 5   B.II.2_1       15870 non-null  int64 \n",
      " 6   C.II.1_1       15870 non-null  int64 \n",
      " 7   C.III.1_1      15870 non-null  int64 \n",
      " 8   HC1_2          15870 non-null  int64 \n",
      " 9   HC2_2          15870 non-null  int64 \n",
      " 10  HC3_2          15870 non-null  int64 \n",
      " 11  B.I.1_2        15870 non-null  int64 \n",
      " 12  B.II.2_2       15870 non-null  int64 \n",
      " 13  C.II.1_2       15870 non-null  int64 \n",
      " 14  C.III.1_2      15870 non-null  int64 \n",
      "dtypes: int64(14), object(1)\n",
      "memory usage: 1.8+ MB\n"
     ]
    }
   ],
   "source": [
    "kuehn_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Converting Ensembl gene IDs into gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper function\n",
    "def get_ensembl_ids(feature, line, remove_dot=False):\n",
    "\n",
    "    '''\n",
    "    Given a line from a genome annotation, returns the requested value.\n",
    "\n",
    "    Arguments:  feature => the value you are interested in extracting, e.g. gene_name\n",
    "                line => line from genome annotation file\n",
    "                remove_dot => when true, removes '.x_y' appended to end of features\n",
    "\n",
    "    Return:    value of the feature you extracted\n",
    "    '''\n",
    "\n",
    "    # some files have e.g. .1_1 appended to ensemble gene id\n",
    "    if remove_dot:\n",
    "        return line.split(feature)[1].split('\"')[1].split('.')[0]\n",
    "\n",
    "    return line.split(feature)[1].split('\"')[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load ucsc annotations\n",
    "with open('../annotations/ensemblToGeneName.txt') as file: # contains ensembl transcript ids and gene names\n",
    "    ensemblToGeneName = list(file)\n",
    "\n",
    "with open('../annotations/hg19.ensGene.gtf') as file: # contains ensembl transcript ids and gene ids\n",
    "    ensGene = list(file)\n",
    "\n",
    "with open('../annotations/ensemblSource.txt') as file:\n",
    "    ensemblSource = list(file)\n",
    "\n",
    "# load gencode comprehensive genome annotation\n",
    "with open('../annotations/gencode.v46lift37.annotation.gtf') as file: # gencode GrCh37 (hg19) comprehensive genome annotation\n",
    "    gencode = list(file)[5:] # remove metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract ensembl transcript ids and corresponding gene ids\n",
    "# NOTE: wanted to do it in this order (transcript_id, gene_id) because transcript_id => gene_id is a surjection\n",
    "#   however, it doesn't actually matter because my data is in gene_ids anyways\n",
    "enst_to_ensg = dict(map(lambda x: (get_ensembl_ids('transcript_id', x), get_ensembl_ids('gene_id', x)), ensGene))\n",
    "\n",
    "# extract ensembl transcript ids and corresponding gene names\n",
    "enst_to_gene = dict(map(lambda x: x.strip('\\n').split('\\t'), ensemblToGeneName))\n",
    "\n",
    "# extract ensembl transcript ids and corresponding gene types\n",
    "ucsc_enst_to_type = dict(map(lambda x: x.strip('\\n').split('\\t'), ensemblSource))\n",
    "\n",
    "# check same length (though they don't need to be)\n",
    "len(enst_to_ensg) == len(enst_to_gene) == len(ucsc_enst_to_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract ensemble gene ids and corresponding gene types\n",
    "gencode_ensg_to_type = dict(\n",
    "    map(lambda x: (get_ensembl_ids('gene_id', x, remove_dot=True), get_ensembl_ids('gene_type', x, remove_dot=True)), \n",
    "        gencode)\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000223972': 'transcribed_unprocessed_pseudogene',\n",
       " 'ENSG00000227232': 'unprocessed_pseudogene',\n",
       " 'ENSG00000243485': 'miRNA',\n",
       " 'ENSG00000237613': 'lincRNA',\n",
       " 'ENSG00000268020': 'unprocessed_pseudogene',\n",
       " 'ENSG00000240361': 'unprocessed_pseudogene',\n",
       " 'ENSG00000186092': 'protein_coding',\n",
       " 'ENSG00000238009': 'lincRNA',\n",
       " 'ENSG00000239945': 'lincRNA',\n",
       " 'ENSG00000233750': 'processed_pseudogene',\n",
       " 'ENSG00000237683': 'protein_coding',\n",
       " 'ENSG00000268903': 'processed_pseudogene',\n",
       " 'ENSG00000269981': 'processed_pseudogene',\n",
       " 'ENSG00000239906': 'antisense',\n",
       " 'ENSG00000241860': 'lincRNA',\n",
       " 'ENSG00000222623': 'snRNA',\n",
       " 'ENSG00000241599': 'lincRNA',\n",
       " 'ENSG00000228463': 'lincRNA',\n",
       " 'ENSG00000241670': 'processed_pseudogene',\n",
       " 'ENSG00000237094': 'lincRNA',\n",
       " 'ENSG00000250575': 'unprocessed_pseudogene',\n",
       " 'ENSG00000233653': 'processed_pseudogene',\n",
       " 'ENSG00000224813': 'transcribed_unprocessed_pseudogene',\n",
       " 'ENSG00000235249': 'protein_coding',\n",
       " 'ENSG00000269732': 'processed_pseudogene',\n",
       " 'ENSG00000256186': 'pseudogene',\n",
       " 'ENSG00000236601': 'lincRNA',\n",
       " 'ENSG00000236743': 'lincRNA',\n",
       " 'ENSG00000236679': 'processed_pseudogene',\n",
       " 'ENSG00000231709': 'lincRNA',\n",
       " 'ENSG00000235146': 'lincRNA',\n",
       " 'ENSG00000239664': 'lincRNA',\n",
       " 'ENSG00000230021': 'lincRNA',\n",
       " 'ENSG00000223659': 'lincRNA',\n",
       " 'ENSG00000225972': 'unprocessed_pseudogene',\n",
       " 'ENSG00000225630': 'unprocessed_pseudogene',\n",
       " 'ENSG00000237973': 'unprocessed_pseudogene',\n",
       " 'ENSG00000229344': 'unprocessed_pseudogene',\n",
       " 'ENSG00000240409': 'unprocessed_pseudogene',\n",
       " 'ENSG00000248527': 'unprocessed_pseudogene',\n",
       " 'ENSG00000198744': 'unprocessed_pseudogene',\n",
       " 'ENSG00000268663': 'processed_pseudogene',\n",
       " 'ENSG00000185097': 'protein_coding',\n",
       " 'ENSG00000229376': 'processed_pseudogene',\n",
       " 'ENSG00000224956': 'unprocessed_pseudogene',\n",
       " 'ENSG00000235373': 'lincRNA',\n",
       " 'ENSG00000223181': 'snRNA',\n",
       " 'ENSG00000240618': 'lincRNA',\n",
       " 'ENSG00000229905': 'antisense',\n",
       " 'ENSG00000228327': 'lincRNA',\n",
       " 'ENSG00000237491': 'lincRNA',\n",
       " 'ENSG00000230092': 'transcribed_unprocessed_pseudogene',\n",
       " 'ENSG00000269831': 'protein_coding',\n",
       " 'ENSG00000240453': 'processed_transcript',\n",
       " 'ENSG00000177757': 'lincRNA',\n",
       " 'ENSG00000225880': 'lincRNA',\n",
       " 'ENSG00000228794': 'retained_intron',\n",
       " 'ENSG00000230368': 'lincRNA',\n",
       " 'ENSG00000234711': 'unprocessed_pseudogene',\n",
       " 'ENSG00000269308': 'protein_coding',\n",
       " 'ENSG00000272438': 'lincRNA',\n",
       " 'ENSG00000230699': 'lincRNA',\n",
       " 'ENSG00000241180': 'lincRNA',\n",
       " 'ENSG00000223764': 'retained_intron',\n",
       " 'ENSG00000187634': 'retained_intron',\n",
       " 'ENSG00000268179': 'protein_coding',\n",
       " 'ENSG00000188976': 'retained_intron',\n",
       " 'ENSG00000187961': 'retained_intron',\n",
       " 'ENSG00000187583': 'protein_coding',\n",
       " 'ENSG00000187642': 'retained_intron',\n",
       " 'ENSG00000272512': 'lincRNA',\n",
       " 'ENSG00000188290': 'protein_coding',\n",
       " 'ENSG00000231702': 'processed_pseudogene',\n",
       " 'ENSG00000224969': 'antisense',\n",
       " 'ENSG00000187608': 'protein_coding',\n",
       " 'ENSG00000188157': 'retained_intron',\n",
       " 'ENSG00000242590': 'sense_intronic',\n",
       " 'ENSG00000217801': 'processed_transcript',\n",
       " 'ENSG00000273443': 'lincRNA',\n",
       " 'ENSG00000237330': 'protein_coding',\n",
       " 'ENSG00000131591': 'protein_coding',\n",
       " 'ENSG00000223823': 'lincRNA',\n",
       " 'ENSG00000207730': 'miRNA',\n",
       " 'ENSG00000207607': 'miRNA',\n",
       " 'ENSG00000198976': 'miRNA',\n",
       " 'ENSG00000272141': 'lincRNA',\n",
       " 'ENSG00000205231': 'antisense',\n",
       " 'ENSG00000162571': 'nonsense_mediated_decay',\n",
       " 'ENSG00000186891': 'protein_coding',\n",
       " 'ENSG00000186827': 'processed_transcript',\n",
       " 'ENSG00000078808': 'processed_transcript',\n",
       " 'ENSG00000176022': 'protein_coding',\n",
       " 'ENSG00000184163': 'retained_intron',\n",
       " 'ENSG00000260179': 'lincRNA',\n",
       " 'ENSG00000160087': 'protein_coding',\n",
       " 'ENSG00000230415': 'lincRNA',\n",
       " 'ENSG00000162572': 'protein_coding',\n",
       " 'ENSG00000131584': 'processed_transcript',\n",
       " 'ENSG00000169972': 'retained_intron',\n",
       " 'ENSG00000127054': 'retained_intron',\n",
       " 'ENSG00000240731': 'sense_intronic',\n",
       " 'ENSG00000224051': 'protein_coding',\n",
       " 'ENSG00000169962': 'protein_coding',\n",
       " 'ENSG00000107404': 'retained_intron',\n",
       " 'ENSG00000162576': 'retained_intron',\n",
       " 'ENSG00000175756': 'processed_transcript',\n",
       " 'ENSG00000223663': 'processed_pseudogene',\n",
       " 'ENSG00000221978': 'protein_coding',\n",
       " 'ENSG00000224870': 'processed_transcript',\n",
       " 'ENSG00000242485': 'processed_transcript',\n",
       " 'ENSG00000264293': 'misc_RNA',\n",
       " 'ENSG00000272455': 'lincRNA',\n",
       " 'ENSG00000235098': 'protein_coding',\n",
       " 'ENSG00000225905': 'antisense',\n",
       " 'ENSG00000205116': 'protein_coding',\n",
       " 'ENSG00000225285': 'lincRNA',\n",
       " 'ENSG00000179403': 'protein_coding',\n",
       " 'ENSG00000215915': 'retained_intron',\n",
       " 'ENSG00000160072': 'retained_intron',\n",
       " 'ENSG00000197785': 'protein_coding',\n",
       " 'ENSG00000205090': 'protein_coding',\n",
       " 'ENSG00000160075': 'protein_coding',\n",
       " 'ENSG00000215791': 'pseudogene',\n",
       " 'ENSG00000215014': 'protein_coding',\n",
       " 'ENSG00000236684': 'pseudogene',\n",
       " 'ENSG00000228594': 'protein_coding',\n",
       " 'ENSG00000272106': 'antisense',\n",
       " 'ENSG00000197530': 'retained_intron',\n",
       " 'ENSG00000189409': 'nonsense_mediated_decay',\n",
       " 'ENSG00000248333': 'protein_coding',\n",
       " 'ENSG00000272004': 'antisense',\n",
       " 'ENSG00000189339': 'retained_intron',\n",
       " 'ENSG00000269737': 'antisense',\n",
       " 'ENSG00000269227': 'transcribed_processed_pseudogene',\n",
       " 'ENSG00000215914': 'unprocessed_pseudogene',\n",
       " 'ENSG00000008128': 'protein_coding',\n",
       " 'ENSG00000268575': 'processed_transcript',\n",
       " 'ENSG00000227775': 'transcribed_processed_pseudogene',\n",
       " 'ENSG00000215790': 'processed_transcript',\n",
       " 'ENSG00000008130': 'processed_transcript',\n",
       " 'ENSG00000078369': 'retained_intron',\n",
       " 'ENSG00000231050': 'antisense',\n",
       " 'ENSG00000169885': 'processed_transcript',\n",
       " 'ENSG00000178821': 'protein_coding',\n",
       " 'ENSG00000142609': 'protein_coding',\n",
       " 'ENSG00000233542': 'antisense',\n",
       " 'ENSG00000187730': 'protein_coding',\n",
       " 'ENSG00000226969': 'antisense',\n",
       " 'ENSG00000067606': 'processed_transcript',\n",
       " 'ENSG00000271806': 'antisense',\n",
       " 'ENSG00000182873': 'antisense',\n",
       " 'ENSG00000162585': 'retained_intron',\n",
       " 'ENSG00000269554': 'protein_coding',\n",
       " 'ENSG00000203301': 'protein_coding',\n",
       " 'ENSG00000243558': 'retained_intron',\n",
       " 'ENSG00000234396': 'lincRNA',\n",
       " 'ENSG00000157933': 'retained_intron',\n",
       " 'ENSG00000116151': 'protein_coding',\n",
       " 'ENSG00000272161': 'sense_intronic',\n",
       " 'ENSG00000269753': 'protein_coding',\n",
       " 'ENSG00000269896': 'processed_transcript',\n",
       " 'ENSG00000238240': 'transcribed_processed_pseudogene',\n",
       " 'ENSG00000272420': 'sense_intronic',\n",
       " 'ENSG00000271921': 'sense_intronic',\n",
       " 'ENSG00000271847': 'sense_intronic',\n",
       " 'ENSG00000178642': 'pseudogene',\n",
       " 'ENSG00000157916': 'retained_intron',\n",
       " 'ENSG00000157911': 'nonsense_mediated_decay',\n",
       " 'ENSG00000149527': 'retained_intron',\n",
       " 'ENSG00000224387': 'antisense',\n",
       " 'ENSG00000229393': 'antisense',\n",
       " 'ENSG00000157881': 'retained_intron',\n",
       " 'ENSG00000197921': 'protein_coding',\n",
       " 'ENSG00000272449': 'lincRNA',\n",
       " 'ENSG00000238164': 'antisense',\n",
       " 'ENSG00000157873': 'processed_transcript',\n",
       " 'ENSG00000225931': 'antisense',\n",
       " 'ENSG00000228037': 'antisense',\n",
       " 'ENSG00000157870': 'processed_transcript',\n",
       " 'ENSG00000142606': 'processed_transcript',\n",
       " 'ENSG00000237058': 'antisense',\n",
       " 'ENSG00000215912': 'protein_coding',\n",
       " 'ENSG00000233234': 'lincRNA',\n",
       " 'ENSG00000231630': 'lincRNA',\n",
       " 'ENSG00000169717': 'protein_coding',\n",
       " 'ENSG00000177133': 'antisense',\n",
       " 'ENSG00000256761': 'pseudogene',\n",
       " 'ENSG00000142611': 'retained_intron',\n",
       " 'ENSG00000226286': 'antisense',\n",
       " 'ENSG00000272235': 'lincRNA',\n",
       " 'ENSG00000130762': 'retained_intron',\n",
       " 'ENSG00000272088': 'lincRNA',\n",
       " 'ENSG00000162591': 'retained_intron',\n",
       " 'ENSG00000207776': 'miRNA',\n",
       " 'ENSG00000238260': 'antisense',\n",
       " 'ENSG00000158109': 'protein_coding',\n",
       " 'ENSG00000116213': 'retained_intron',\n",
       " 'ENSG00000078900': 'retained_intron',\n",
       " 'ENSG00000227589': 'antisense',\n",
       " 'ENSG00000235131': 'antisense',\n",
       " 'ENSG00000227372': 'antisense',\n",
       " 'ENSG00000162592': 'nonsense_mediated_decay',\n",
       " 'ENSG00000235169': 'protein_coding',\n",
       " 'ENSG00000130764': 'nonsense_mediated_decay',\n",
       " 'ENSG00000266075': 'misc_RNA',\n",
       " 'ENSG00000272153': 'antisense',\n",
       " 'ENSG00000116198': 'protein_coding',\n",
       " 'ENSG00000169598': 'nonsense_mediated_decay',\n",
       " 'ENSG00000264428': 'miRNA',\n",
       " 'ENSG00000198912': 'processed_transcript',\n",
       " 'ENSG00000236423': 'lincRNA',\n",
       " 'ENSG00000233304': 'lincRNA',\n",
       " 'ENSG00000226374': 'lincRNA',\n",
       " 'ENSG00000229280': 'processed_pseudogene',\n",
       " 'ENSG00000235054': 'lincRNA',\n",
       " 'ENSG00000227169': 'lincRNA',\n",
       " 'ENSG00000232596': 'lincRNA',\n",
       " 'ENSG00000196581': 'processed_transcript',\n",
       " 'ENSG00000231510': 'lincRNA',\n",
       " 'ENSG00000260972': 'lincRNA',\n",
       " 'ENSG00000264341': 'miRNA',\n",
       " 'ENSG00000236948': 'lincRNA',\n",
       " 'ENSG00000216045': 'miRNA',\n",
       " 'ENSG00000264101': 'miRNA',\n",
       " 'ENSG00000131697': 'nonsense_mediated_decay',\n",
       " 'ENSG00000266687': 'miRNA',\n",
       " 'ENSG00000069424': 'retained_intron',\n",
       " 'ENSG00000268332': 'protein_coding',\n",
       " 'ENSG00000116254': 'nonsense_mediated_decay',\n",
       " 'ENSG00000116251': 'protein_coding',\n",
       " 'ENSG00000226944': 'antisense',\n",
       " 'ENSG00000158286': 'processed_transcript',\n",
       " 'ENSG00000116237': 'protein_coding',\n",
       " 'ENSG00000225077': 'antisense',\n",
       " 'ENSG00000173673': 'protein_coding',\n",
       " 'ENSG00000158292': 'protein_coding',\n",
       " 'ENSG00000097021': 'protein_coding',\n",
       " 'ENSG00000271746': 'antisense',\n",
       " 'ENSG00000069812': 'protein_coding',\n",
       " 'ENSG00000187017': 'processed_transcript',\n",
       " 'ENSG00000265392': 'miRNA',\n",
       " 'ENSG00000231868': 'antisense',\n",
       " 'ENSG00000215788': 'retained_intron',\n",
       " 'ENSG00000171680': 'protein_coding',\n",
       " 'ENSG00000162408': 'retained_intron',\n",
       " 'ENSG00000253022': 'snRNA',\n",
       " 'ENSG00000229519': 'processed_pseudogene',\n",
       " 'ENSG00000173662': 'protein_coding',\n",
       " 'ENSG00000204859': 'retained_intron',\n",
       " 'ENSG00000162413': 'protein_coding',\n",
       " 'ENSG00000116273': 'processed_transcript',\n",
       " 'ENSG00000041988': 'retained_intron',\n",
       " 'ENSG00000007923': 'processed_transcript',\n",
       " 'ENSG00000228750': 'lincRNA',\n",
       " 'ENSG00000239166': 'snoRNA',\n",
       " 'ENSG00000237436': 'antisense',\n",
       " 'ENSG00000171735': 'protein_coding',\n",
       " 'ENSG00000227950': 'processed_pseudogene',\n",
       " 'ENSG00000237365': 'antisense',\n",
       " 'ENSG00000207056': 'snRNA',\n",
       " 'ENSG00000237402': 'sense_intronic',\n",
       " 'ENSG00000237728': 'antisense',\n",
       " 'ENSG00000225126': 'antisense',\n",
       " 'ENSG00000270171': 'lincRNA',\n",
       " 'ENSG00000270035': 'lincRNA',\n",
       " 'ENSG00000269978': 'lincRNA',\n",
       " 'ENSG00000049245': 'retained_intron',\n",
       " 'ENSG00000269925': 'sense_intronic',\n",
       " 'ENSG00000049246': 'retained_intron',\n",
       " 'ENSG00000236266': 'antisense',\n",
       " 'ENSG00000049247': 'protein_coding',\n",
       " 'ENSG00000049249': 'nonsense_mediated_decay',\n",
       " 'ENSG00000270330': 'processed_pseudogene',\n",
       " 'ENSG00000116288': 'protein_coding',\n",
       " 'ENSG00000200344': 'misc_RNA',\n",
       " 'ENSG00000116285': 'nonsense_mediated_decay',\n",
       " 'ENSG00000238290': 'lincRNA',\n",
       " 'ENSG00000232848': 'processed_pseudogene',\n",
       " 'ENSG00000229305': 'processed_pseudogene',\n",
       " 'ENSG00000200975': 'snRNA',\n",
       " 'ENSG00000227634': 'lincRNA',\n",
       " 'ENSG00000233645': 'lincRNA',\n",
       " 'ENSG00000265414': 'misc_RNA',\n",
       " 'ENSG00000251977': 'snRNA',\n",
       " 'ENSG00000162426': 'processed_transcript',\n",
       " 'ENSG00000252254': 'misc_RNA',\n",
       " 'ENSG00000142599': 'protein_coding',\n",
       " 'ENSG00000232912': 'antisense',\n",
       " 'ENSG00000221083': 'snoRNA',\n",
       " 'ENSG00000270282': 'processed_pseudogene',\n",
       " 'ENSG00000234619': 'processed_pseudogene',\n",
       " 'ENSG00000224315': 'processed_pseudogene',\n",
       " 'ENSG00000222511': 'misc_RNA',\n",
       " 'ENSG00000226545': 'unprocessed_pseudogene',\n",
       " 'ENSG00000228423': 'lincRNA',\n",
       " 'ENSG00000229716': 'processed_pseudogene',\n",
       " 'ENSG00000074800': 'protein_coding',\n",
       " 'ENSG00000236269': 'sense_intronic',\n",
       " 'ENSG00000230679': 'antisense',\n",
       " 'ENSG00000201725': 'snRNA',\n",
       " 'ENSG00000238249': 'processed_pseudogene',\n",
       " 'ENSG00000232208': 'processed_pseudogene',\n",
       " 'ENSG00000232663': 'processed_pseudogene',\n",
       " 'ENSG00000131686': 'protein_coding',\n",
       " 'ENSG00000265141': 'misc_RNA',\n",
       " 'ENSG00000197241': 'protein_coding',\n",
       " 'ENSG00000142583': 'protein_coding',\n",
       " 'ENSG00000252404': 'snoRNA',\n",
       " 'ENSG00000180758': 'retained_intron',\n",
       " 'ENSG00000207865': 'miRNA',\n",
       " 'ENSG00000228526': 'lincRNA',\n",
       " 'ENSG00000234546': 'lincRNA',\n",
       " 'ENSG00000049239': 'processed_transcript',\n",
       " 'ENSG00000252841': 'miRNA',\n",
       " 'ENSG00000171621': 'protein_coding',\n",
       " 'ENSG00000252956': 'rRNA',\n",
       " 'ENSG00000235263': 'lincRNA',\n",
       " 'ENSG00000171612': 'protein_coding',\n",
       " 'ENSG00000231181': 'processed_pseudogene',\n",
       " 'ENSG00000188807': 'nonsense_mediated_decay',\n",
       " 'ENSG00000171608': 'protein_coding',\n",
       " 'ENSG00000179840': 'protein_coding',\n",
       " 'ENSG00000233268': 'processed_pseudogene',\n",
       " 'ENSG00000231789': 'antisense',\n",
       " 'ENSG00000171603': 'processed_transcript',\n",
       " 'ENSG00000271264': 'processed_pseudogene',\n",
       " 'ENSG00000178585': 'protein_coding',\n",
       " 'ENSG00000223989': 'antisense',\n",
       " 'ENSG00000162441': 'retained_intron',\n",
       " 'ENSG00000228150': 'sense_intronic',\n",
       " 'ENSG00000173614': 'protein_coding',\n",
       " 'ENSG00000202415': 'misc_RNA',\n",
       " 'ENSG00000265521': 'miRNA',\n",
       " 'ENSG00000241326': 'sense_intronic',\n",
       " 'ENSG00000162444': 'protein_coding',\n",
       " 'ENSG00000225786': 'pseudogene',\n",
       " 'ENSG00000130939': 'processed_transcript',\n",
       " 'ENSG00000224340': 'processed_pseudogene',\n",
       " 'ENSG00000233623': 'processed_pseudogene',\n",
       " 'ENSG00000201746': 'snRNA',\n",
       " 'ENSG00000054523': 'retained_intron',\n",
       " 'ENSG00000264881': 'miRNA',\n",
       " 'ENSG00000199562': 'snRNA',\n",
       " 'ENSG00000264501': 'misc_RNA',\n",
       " 'ENSG00000265945': 'misc_RNA',\n",
       " 'ENSG00000142657': 'retained_intron',\n",
       " 'ENSG00000271989': 'antisense',\n",
       " 'ENSG00000251503': 'protein_coding',\n",
       " 'ENSG00000175279': 'retained_intron',\n",
       " 'ENSG00000241563': 'protein_coding',\n",
       " 'ENSG00000160049': 'protein_coding',\n",
       " 'ENSG00000203469': 'antisense',\n",
       " 'ENSG00000142655': 'processed_transcript',\n",
       " 'ENSG00000243267': 'misc_RNA',\n",
       " 'ENSG00000130940': 'processed_transcript',\n",
       " 'ENSG00000272078': 'antisense',\n",
       " 'ENSG00000225158': 'processed_pseudogene',\n",
       " 'ENSG00000175262': 'protein_coding',\n",
       " 'ENSG00000215785': 'processed_pseudogene',\n",
       " 'ENSG00000264181': 'miRNA',\n",
       " 'ENSG00000207213': 'misc_RNA',\n",
       " 'ENSG00000120948': 'processed_transcript',\n",
       " 'ENSG00000009724': 'retained_intron',\n",
       " 'ENSG00000271895': 'antisense',\n",
       " 'ENSG00000116649': 'protein_coding',\n",
       " 'ENSG00000171824': 'nonsense_mediated_decay',\n",
       " 'ENSG00000226849': 'antisense',\n",
       " 'ENSG00000230337': 'antisense',\n",
       " 'ENSG00000198793': 'processed_transcript',\n",
       " 'ENSG00000225602': 'antisense',\n",
       " 'ENSG00000253086': 'snRNA',\n",
       " 'ENSG00000271794': 'snoRNA',\n",
       " 'ENSG00000171819': 'processed_transcript',\n",
       " 'ENSG00000207451': 'snRNA',\n",
       " 'ENSG00000238173': 'processed_pseudogene',\n",
       " 'ENSG00000120942': 'protein_coding',\n",
       " 'ENSG00000238199': 'processed_pseudogene',\n",
       " 'ENSG00000224338': 'processed_pseudogene',\n",
       " 'ENSG00000204624': 'protein_coding',\n",
       " 'ENSG00000235643': 'lincRNA',\n",
       " 'ENSG00000116661': 'processed_transcript',\n",
       " 'ENSG00000132879': 'protein_coding',\n",
       " 'ENSG00000116663': 'protein_coding',\n",
       " 'ENSG00000116670': 'retained_intron',\n",
       " 'ENSG00000162490': 'protein_coding',\n",
       " 'ENSG00000177674': 'processed_transcript',\n",
       " 'ENSG00000215910': 'processed_transcript',\n",
       " 'ENSG00000177553': 'antisense',\n",
       " 'ENSG00000177000': 'protein_coding',\n",
       " 'ENSG00000011021': 'processed_transcript',\n",
       " 'ENSG00000242349': 'antisense',\n",
       " 'ENSG00000175206': 'protein_coding',\n",
       " 'ENSG00000120937': 'protein_coding',\n",
       " 'ENSG00000224904': 'processed_pseudogene',\n",
       " 'ENSG00000199347': 'snRNA',\n",
       " 'ENSG00000201801': 'snRNA',\n",
       " 'ENSG00000116685': 'protein_coding',\n",
       " 'ENSG00000083444': 'retained_intron',\n",
       " 'ENSG00000116688': 'protein_coding',\n",
       " 'ENSG00000270914': 'processed_pseudogene',\n",
       " 'ENSG00000116691': 'processed_transcript',\n",
       " 'ENSG00000202461': 'misc_RNA',\n",
       " 'ENSG00000243539': 'misc_RNA',\n",
       " 'ENSG00000120949': 'processed_transcript',\n",
       " 'ENSG00000201135': 'snRNA',\n",
       " 'ENSG00000227875': 'processed_pseudogene',\n",
       " 'ENSG00000028137': 'processed_transcript',\n",
       " 'ENSG00000263676': 'miRNA',\n",
       " 'ENSG00000225196': 'processed_pseudogene',\n",
       " 'ENSG00000252969': 'snoRNA',\n",
       " 'ENSG00000048707': 'retained_intron',\n",
       " 'ENSG00000239149': 'snoRNA',\n",
       " 'ENSG00000229484': 'lincRNA',\n",
       " 'ENSG00000162496': 'processed_transcript',\n",
       " 'ENSG00000221340': 'snRNA',\n",
       " 'ENSG00000272482': 'lincRNA',\n",
       " 'ENSG00000204518': 'protein_coding',\n",
       " 'ENSG00000226166': 'processed_pseudogene',\n",
       " 'ENSG00000188984': 'protein_coding',\n",
       " 'ENSG00000238771': 'snoRNA',\n",
       " 'ENSG00000157330': 'protein_coding',\n",
       " 'ENSG00000116726': 'protein_coding',\n",
       " 'ENSG00000116721': 'protein_coding',\n",
       " 'ENSG00000228338': 'lincRNA',\n",
       " 'ENSG00000204513': 'protein_coding',\n",
       " 'ENSG00000231103': 'unprocessed_pseudogene',\n",
       " 'ENSG00000179172': 'protein_coding',\n",
       " 'ENSG00000120952': 'protein_coding',\n",
       " 'ENSG00000243073': 'protein_coding',\n",
       " 'ENSG00000187545': 'protein_coding',\n",
       " 'ENSG00000204510': 'protein_coding',\n",
       " 'ENSG00000207434': 'snRNA',\n",
       " 'ENSG00000234064': 'unprocessed_pseudogene',\n",
       " 'ENSG00000232423': 'protein_coding',\n",
       " 'ENSG00000250894': 'unprocessed_pseudogene',\n",
       " 'ENSG00000204508': 'protein_coding',\n",
       " 'ENSG00000239810': 'protein_coding',\n",
       " 'ENSG00000270601': 'unprocessed_pseudogene',\n",
       " 'ENSG00000250113': 'unprocessed_pseudogene',\n",
       " 'ENSG00000204505': 'unprocessed_pseudogene',\n",
       " 'ENSG00000233336': 'lincRNA',\n",
       " 'ENSG00000229978': 'processed_transcript',\n",
       " 'ENSG00000236179': 'unprocessed_pseudogene',\n",
       " 'ENSG00000179412': 'processed_pseudogene',\n",
       " 'ENSG00000224199': 'unprocessed_pseudogene',\n",
       " 'ENSG00000229571': 'protein_coding',\n",
       " 'ENSG00000204503': 'protein_coding',\n",
       " 'ENSG00000233394': 'unprocessed_pseudogene',\n",
       " 'ENSG00000204502': 'protein_coding',\n",
       " 'ENSG00000231219': 'unprocessed_pseudogene',\n",
       " 'ENSG00000207511': 'snRNA',\n",
       " 'ENSG00000182330': 'protein_coding',\n",
       " 'ENSG00000237700': 'unitary_pseudogene',\n",
       " 'ENSG00000204501': 'protein_coding',\n",
       " 'ENSG00000228229': 'lincRNA',\n",
       " 'ENSG00000204495': 'protein_coding',\n",
       " 'ENSG00000204491': 'protein_coding',\n",
       " 'ENSG00000204488': 'protein_coding',\n",
       " 'ENSG00000204486': 'protein_coding',\n",
       " 'ENSG00000204485': 'unprocessed_pseudogene',\n",
       " 'ENSG00000224581': 'unprocessed_pseudogene',\n",
       " 'ENSG00000157358': 'protein_coding',\n",
       " 'ENSG00000204481': 'protein_coding',\n",
       " 'ENSG00000204480': 'protein_coding',\n",
       " 'ENSG00000204479': 'protein_coding',\n",
       " 'ENSG00000204478': 'protein_coding',\n",
       " 'ENSG00000162494': 'protein_coding',\n",
       " 'ENSG00000259961': 'antisense',\n",
       " 'ENSG00000215909': 'processed_pseudogene',\n",
       " 'ENSG00000162493': 'nonsense_mediated_decay',\n",
       " 'ENSG00000222952': 'rRNA',\n",
       " 'ENSG00000266358': 'miRNA',\n",
       " 'ENSG00000237445': 'lincRNA',\n",
       " 'ENSG00000253085': 'snoRNA',\n",
       " 'ENSG00000116731': 'protein_coding',\n",
       " 'ENSG00000252151': 'snRNA',\n",
       " 'ENSG00000231606': 'lincRNA',\n",
       " 'ENSG00000234593': 'lincRNA',\n",
       " 'ENSG00000189337': 'protein_coding',\n",
       " 'ENSG00000235829': 'processed_pseudogene',\n",
       " 'ENSG00000175147': 'processed_transcript',\n",
       " 'ENSG00000171729': 'protein_coding',\n",
       " 'ENSG00000204464': 'protein_coding',\n",
       " 'ENSG00000215720': 'processed_pseudogene',\n",
       " 'ENSG00000142621': 'processed_transcript',\n",
       " 'ENSG00000233485': 'antisense',\n",
       " 'ENSG00000236045': 'antisense',\n",
       " 'ENSG00000228140': 'lincRNA',\n",
       " 'ENSG00000142634': 'protein_coding',\n",
       " 'ENSG00000162438': 'processed_transcript',\n",
       " 'ENSG00000142615': 'processed_transcript',\n",
       " 'ENSG00000215704': 'processed_transcript',\n",
       " 'ENSG00000132906': 'processed_transcript',\n",
       " 'ENSG00000116138': 'processed_transcript',\n",
       " 'ENSG00000251866': 'snoRNA',\n",
       " 'ENSG00000272510': 'antisense',\n",
       " 'ENSG00000116771': 'protein_coding',\n",
       " 'ENSG00000237301': 'antisense',\n",
       " 'ENSG00000235084': 'processed_pseudogene',\n",
       " 'ENSG00000252417': 'snRNA',\n",
       " 'ENSG00000236500': 'processed_pseudogene',\n",
       " 'ENSG00000197312': 'retained_intron',\n",
       " 'ENSG00000215695': 'protein_coding',\n",
       " 'ENSG00000271742': 'lincRNA',\n",
       " 'ENSG00000116786': 'processed_transcript',\n",
       " 'ENSG00000264048': 'miRNA',\n",
       " 'ENSG00000237938': 'antisense',\n",
       " 'ENSG00000162461': 'processed_transcript',\n",
       " 'ENSG00000224459': 'antisense',\n",
       " 'ENSG00000162460': 'processed_transcript',\n",
       " 'ENSG00000162458': 'processed_transcript',\n",
       " 'ENSG00000224321': 'processed_pseudogene',\n",
       " 'ENSG00000233954': 'processed_pseudogene',\n",
       " 'ENSG00000178715': 'processed_pseudogene',\n",
       " 'ENSG00000179743': 'antisense',\n",
       " 'ENSG00000065526': 'processed_transcript',\n",
       " 'ENSG00000238818': 'snoRNA',\n",
       " 'ENSG00000270620': 'unprocessed_pseudogene',\n",
       " 'ENSG00000116809': 'processed_transcript',\n",
       " 'ENSG00000234607': 'processed_pseudogene',\n",
       " 'ENSG00000231353': 'unprocessed_pseudogene',\n",
       " 'ENSG00000224966': 'unprocessed_pseudogene',\n",
       " 'ENSG00000183888': 'protein_coding',\n",
       " 'ENSG00000233078': 'antisense',\n",
       " 'ENSG00000173641': 'protein_coding',\n",
       " 'ENSG00000186510': 'processed_transcript',\n",
       " 'ENSG00000232456': 'unprocessed_pseudogene',\n",
       " 'ENSG00000184908': 'protein_coding',\n",
       " 'ENSG00000185519': 'processed_transcript',\n",
       " 'ENSG00000142627': 'processed_transcript',\n",
       " 'ENSG00000227959': 'antisense',\n",
       " 'ENSG00000224621': 'antisense',\n",
       " 'ENSG00000142632': 'protein_coding',\n",
       " 'ENSG00000234166': 'antisense',\n",
       " 'ENSG00000237276': 'transcribed_unprocessed_pseudogene',\n",
       " 'ENSG00000204377': 'protein_coding',\n",
       " 'ENSG00000132881': 'protein_coding',\n",
       " 'ENSG00000233929': 'processed_pseudogene',\n",
       " 'ENSG00000037637': 'processed_transcript',\n",
       " 'ENSG00000055070': 'processed_transcript',\n",
       " 'ENSG00000226457': 'processed_pseudogene',\n",
       " 'ENSG00000187144': 'protein_coding',\n",
       " 'ENSG00000157191': 'retained_intron',\n",
       " 'ENSG00000226029': 'lincRNA',\n",
       " 'ENSG00000080947': 'processed_transcript',\n",
       " 'ENSG00000206652': 'snRNA',\n",
       " 'ENSG00000261135': 'lincRNA',\n",
       " 'ENSG00000224174': 'lincRNA',\n",
       " 'ENSG00000233421': 'snRNA',\n",
       " 'ENSG00000268674': 'protein_coding',\n",
       " 'ENSG00000225710': 'unprocessed_pseudogene',\n",
       " 'ENSG00000266584': 'miRNA',\n",
       " 'ENSG00000219481': 'processed_transcript',\n",
       " 'ENSG00000271732': 'lincRNA',\n",
       " 'ENSG00000215908': 'retained_intron',\n",
       " 'ENSG00000186301': 'unprocessed_pseudogene',\n",
       " 'ENSG00000207513': 'snRNA',\n",
       " 'ENSG00000237847': 'protein_coding',\n",
       " 'ENSG00000264742': 'miRNA',\n",
       " 'ENSG00000236698': 'processed_pseudogene',\n",
       " 'ENSG00000268869': 'transcribed_unprocessed_pseudogene',\n",
       " 'ENSG00000268991': 'protein_coding',\n",
       " 'ENSG00000058453': 'processed_transcript',\n",
       " 'ENSG00000207389': 'snRNA',\n",
       " 'ENSG00000227684': 'unprocessed_pseudogene',\n",
       " 'ENSG00000186715': 'processed_transcript',\n",
       " 'ENSG00000223643': 'lincRNA',\n",
       " 'ENSG00000230239': 'lincRNA',\n",
       " 'ENSG00000263811': 'miRNA',\n",
       " 'ENSG00000228549': 'snRNA',\n",
       " 'ENSG00000238142': 'lincRNA',\n",
       " 'ENSG00000235241': 'processed_pseudogene',\n",
       " 'ENSG00000207005': 'snRNA',\n",
       " 'ENSG00000272426': 'lincRNA',\n",
       " 'ENSG00000117122': 'processed_transcript',\n",
       " 'ENSG00000226526': 'antisense',\n",
       " 'ENSG00000159363': 'protein_coding',\n",
       " 'ENSG00000117118': 'processed_transcript',\n",
       " 'ENSG00000117115': 'protein_coding',\n",
       " 'ENSG00000252180': 'misc_RNA',\n",
       " 'ENSG00000204362': 'lincRNA',\n",
       " 'ENSG00000142623': 'processed_transcript',\n",
       " 'ENSG00000142619': 'protein_coding',\n",
       " 'ENSG00000266634': 'miRNA',\n",
       " 'ENSG00000159339': 'processed_transcript',\n",
       " 'ENSG00000269468': 'protein_coding',\n",
       " 'ENSG00000256049': 'processed_transcript',\n",
       " 'ENSG00000179051': 'protein_coding',\n",
       " 'ENSG00000227751': 'antisense',\n",
       " 'ENSG00000266727': 'miRNA',\n",
       " 'ENSG00000239020': 'snoRNA',\n",
       " 'ENSG00000074964': 'processed_transcript',\n",
       " 'ENSG00000236648': 'lincRNA',\n",
       " 'ENSG00000117148': 'protein_coding',\n",
       " 'ENSG00000261781': 'lincRNA',\n",
       " 'ENSG00000117154': 'processed_transcript',\n",
       " 'ENSG00000230035': 'antisense',\n",
       " 'ENSG00000225387': 'lincRNA',\n",
       " 'ENSG00000179023': 'protein_coding',\n",
       " 'ENSG00000235282': 'processed_pseudogene',\n",
       " 'ENSG00000225478': 'processed_pseudogene',\n",
       " 'ENSG00000009709': 'protein_coding',\n",
       " 'ENSG00000179002': 'protein_coding',\n",
       " 'ENSG00000255275': 'nonsense_mediated_decay',\n",
       " 'ENSG00000159423': 'protein_coding',\n",
       " 'ENSG00000265606': 'miRNA',\n",
       " 'ENSG00000221662': 'miRNA',\n",
       " 'ENSG00000169991': 'protein_coding',\n",
       " 'ENSG00000272084': '3prime_overlapping_ncrna',\n",
       " 'ENSG00000127481': 'processed_transcript',\n",
       " 'ENSG00000266098': 'miRNA',\n",
       " 'ENSG00000230424': 'antisense',\n",
       " 'ENSG00000127463': 'protein_coding',\n",
       " 'ENSG00000053372': 'processed_transcript',\n",
       " 'ENSG00000211454': 'retained_intron',\n",
       " 'ENSG00000162482': 'protein_coding',\n",
       " 'ENSG00000270728': 'processed_pseudogene',\n",
       " 'ENSG00000053371': 'retained_intron',\n",
       " 'ENSG00000200403': 'snRNA',\n",
       " 'ENSG00000040487': 'protein_coding',\n",
       " 'ENSG00000265192': 'misc_RNA',\n",
       " 'ENSG00000077549': 'processed_transcript',\n",
       " 'ENSG00000240490': 'misc_RNA',\n",
       " 'ENSG00000201609': 'snRNA',\n",
       " 'ENSG00000239027': 'snoRNA',\n",
       " 'ENSG00000235185': 'antisense',\n",
       " 'ENSG00000270136': 'nonsense_mediated_decay',\n",
       " 'ENSG00000173436': 'processed_transcript',\n",
       " 'ENSG00000226396': 'processed_pseudogene',\n",
       " 'ENSG00000158747': 'protein_coding',\n",
       " 'ENSG00000158748': 'protein_coding',\n",
       " 'ENSG00000162542': 'processed_transcript',\n",
       " 'ENSG00000178828': 'protein_coding',\n",
       " 'ENSG00000235434': 'antisense',\n",
       " 'ENSG00000169914': 'processed_transcript',\n",
       " 'ENSG00000188784': 'protein_coding',\n",
       " 'ENSG00000242688': 'misc_RNA',\n",
       " 'ENSG00000188257': 'processed_transcript',\n",
       " 'ENSG00000127472': 'processed_transcript',\n",
       " 'ENSG00000117215': 'protein_coding',\n",
       " 'ENSG00000158786': 'retained_intron',\n",
       " 'ENSG00000227066': 'lincRNA',\n",
       " 'ENSG00000187980': 'protein_coding',\n",
       " 'ENSG00000225986': 'antisense',\n",
       " 'ENSG00000162543': 'protein_coding',\n",
       " 'ENSG00000228105': 'lincRNA',\n",
       " 'ENSG00000158816': 'processed_transcript',\n",
       " 'ENSG00000226664': 'antisense',\n",
       " 'ENSG00000236963': 'lincRNA',\n",
       " 'ENSG00000226487': 'lincRNA',\n",
       " 'ENSG00000162545': 'processed_transcript',\n",
       " 'ENSG00000090432': 'protein_coding',\n",
       " 'ENSG00000215520': 'processed_pseudogene',\n",
       " 'ENSG00000183114': 'protein_coding',\n",
       " 'ENSG00000158825': 'processed_transcript',\n",
       " 'ENSG00000158828': 'processed_transcript',\n",
       " 'ENSG00000117242': 'retained_intron',\n",
       " 'ENSG00000257033': 'pseudogene',\n",
       " 'ENSG00000244038': 'processed_transcript',\n",
       " 'ENSG00000117245': 'processed_transcript',\n",
       " 'ENSG00000235432': 'unprocessed_pseudogene',\n",
       " 'ENSG00000189410': 'protein_coding',\n",
       " 'ENSG00000233069': 'antisense',\n",
       " 'ENSG00000127483': 'protein_coding',\n",
       " 'ENSG00000203394': 'antisense',\n",
       " 'ENSG00000075151': 'processed_transcript',\n",
       " 'ENSG00000251914': 'snRNA',\n",
       " 'ENSG00000221808': 'miRNA',\n",
       " 'ENSG00000233072': 'processed_pseudogene',\n",
       " 'ENSG00000236073': 'processed_pseudogene',\n",
       " 'ENSG00000235112': 'processed_pseudogene',\n",
       " 'ENSG00000266867': 'miRNA',\n",
       " 'ENSG00000117298': 'processed_transcript',\n",
       " 'ENSG00000236936': 'antisense',\n",
       " 'ENSG00000231105': 'antisense',\n",
       " 'ENSG00000236216': 'processed_pseudogene',\n",
       " 'ENSG00000228823': 'unprocessed_pseudogene',\n",
       " 'ENSG00000236009': 'lincRNA',\n",
       " 'ENSG00000227001': 'unprocessed_pseudogene',\n",
       " 'ENSG00000187952': 'processed_pseudogene',\n",
       " 'ENSG00000186543': 'unprocessed_pseudogene',\n",
       " 'ENSG00000142794': 'processed_transcript',\n",
       " 'ENSG00000176378': 'processed_pseudogene',\n",
       " 'ENSG00000162551': 'protein_coding',\n",
       " 'ENSG00000233431': 'antisense',\n",
       " 'ENSG00000221781': 'miRNA',\n",
       " 'ENSG00000076864': 'processed_transcript',\n",
       " 'ENSG00000090686': 'processed_transcript',\n",
       " 'ENSG00000231978': 'processed_pseudogene',\n",
       " 'ENSG00000187942': 'processed_transcript',\n",
       " 'ENSG00000142798': 'protein_coding',\n",
       " 'ENSG00000232037': 'processed_pseudogene',\n",
       " 'ENSG00000219073': 'processed_transcript',\n",
       " 'ENSG00000264385': 'misc_RNA',\n",
       " 'ENSG00000263991': 'misc_RNA',\n",
       " 'ENSG00000201919': 'snRNA',\n",
       " 'ENSG00000142789': 'protein_coding',\n",
       " 'ENSG00000263582': 'misc_RNA',\n",
       " 'ENSG00000263929': 'misc_RNA',\n",
       " 'ENSG00000201273': 'snRNA',\n",
       " 'ENSG00000228397': 'lincRNA',\n",
       " 'ENSG00000218510': 'lincRNA',\n",
       " 'ENSG00000070831': 'protein_coding',\n",
       " 'ENSG00000230068': 'sense_intronic',\n",
       " 'ENSG00000271428': 'processed_pseudogene',\n",
       " 'ENSG00000271840': 'lincRNA',\n",
       " 'ENSG00000162552': 'protein_coding',\n",
       " 'ENSG00000266564': 'miRNA',\n",
       " 'ENSG00000234397': 'processed_pseudogene',\n",
       " 'ENSG00000184677': 'protein_coding',\n",
       " 'ENSG00000237200': 'sense_intronic',\n",
       " 'ENSG00000070886': 'protein_coding',\n",
       " 'ENSG00000173372': 'protein_coding',\n",
       " 'ENSG00000159189': 'protein_coding',\n",
       " 'ENSG00000173369': 'protein_coding',\n",
       " 'ENSG00000133216': 'processed_transcript',\n",
       " 'ENSG00000265422': 'miRNA',\n",
       " 'ENSG00000216157': 'miRNA',\n",
       " 'ENSG00000225952': 'antisense',\n",
       " 'ENSG00000264014': 'miRNA',\n",
       " 'ENSG00000272796': 'lincRNA',\n",
       " 'ENSG00000215906': 'protein_coding',\n",
       " 'ENSG00000227868': 'protein_coding',\n",
       " 'ENSG00000004487': 'retained_intron',\n",
       " 'ENSG00000240553': 'antisense',\n",
       " 'ENSG00000263793': 'miRNA',\n",
       " 'ENSG00000266802': 'miRNA',\n",
       " 'ENSG00000169641': 'protein_coding',\n",
       " 'ENSG00000229010': 'processed_pseudogene',\n",
       " 'ENSG00000206935': 'snRNA',\n",
       " 'ENSG00000252578': 'snRNA',\n",
       " 'ENSG00000179546': 'protein_coding',\n",
       " 'ENSG00000215381': 'processed_pseudogene',\n",
       " 'ENSG00000261326': 'lincRNA',\n",
       " 'ENSG00000125944': 'processed_transcript',\n",
       " 'ENSG00000125945': 'protein_coding',\n",
       " 'ENSG00000249087': 'protein_coding',\n",
       " 'ENSG00000201405': 'misc_RNA',\n",
       " 'ENSG00000271420': 'lincRNA',\n",
       " 'ENSG00000204219': 'retained_intron',\n",
       " 'ENSG00000232482': 'processed_pseudogene',\n",
       " 'ENSG00000088280': 'processed_transcript',\n",
       " 'ENSG00000007968': 'processed_transcript',\n",
       " 'ENSG00000235052': 'lincRNA',\n",
       " 'ENSG00000117318': 'processed_transcript',\n",
       " 'ENSG00000197880': 'protein_coding',\n",
       " 'ENSG00000229239': 'unprocessed_pseudogene',\n",
       " 'ENSG00000142676': 'protein_coding',\n",
       " 'ENSG00000203589': 'processed_pseudogene',\n",
       " 'ENSG00000244609': 'misc_RNA',\n",
       " 'ENSG00000011007': 'processed_transcript',\n",
       " 'ENSG00000236810': 'antisense',\n",
       " 'ENSG00000057757': 'protein_coding',\n",
       " 'ENSG00000011009': 'processed_transcript',\n",
       " 'ENSG00000117308': 'processed_transcript',\n",
       " 'ENSG00000117305': 'processed_transcript',\n",
       " 'ENSG00000179163': 'protein_coding',\n",
       " 'ENSG00000188822': 'protein_coding',\n",
       " 'ENSG00000264347': 'misc_RNA',\n",
       " 'ENSG00000238538': 'snoRNA',\n",
       " 'ENSG00000229106': 'processed_pseudogene',\n",
       " 'ENSG00000232557': 'antisense',\n",
       " 'ENSG00000264926': 'miRNA',\n",
       " 'ENSG00000230228': 'processed_pseudogene',\n",
       " 'ENSG00000189266': 'protein_coding',\n",
       " 'ENSG00000188529': 'retained_intron',\n",
       " 'ENSG00000215905': 'processed_pseudogene',\n",
       " 'ENSG00000225315': 'antisense',\n",
       " 'ENSG00000142661': 'processed_transcript',\n",
       " 'ENSG00000230703': 'antisense',\n",
       " 'ENSG00000142677': 'protein_coding',\n",
       " 'ENSG00000185436': 'protein_coding',\n",
       " 'ENSG00000230023': 'lincRNA',\n",
       " 'ENSG00000221032': 'miRNA',\n",
       " 'ENSG00000266511': 'miRNA',\n",
       " 'ENSG00000232298': 'antisense',\n",
       " 'ENSG00000158055': 'retained_intron',\n",
       " 'ENSG00000001460': 'processed_transcript',\n",
       " 'ENSG00000001461': 'protein_coding',\n",
       " 'ENSG00000117602': 'nonsense_mediated_decay',\n",
       " 'ENSG00000266551': 'misc_RNA',\n",
       " 'ENSG00000264443': 'lincRNA',\n",
       " 'ENSG00000184454': 'processed_transcript',\n",
       " 'ENSG00000227312': 'processed_pseudogene',\n",
       " 'ENSG00000239106': 'misc_RNA',\n",
       " 'ENSG00000133226': 'retained_intron',\n",
       " 'ENSG00000238986': 'snoRNA',\n",
       " 'ENSG00000169504': 'processed_transcript',\n",
       " 'ENSG00000238482': 'snRNA',\n",
       " 'ENSG00000020633': 'processed_transcript',\n",
       " 'ENSG00000229162': 'antisense',\n",
       " 'ENSG00000261025': 'lincRNA',\n",
       " 'ENSG00000264371': 'miRNA',\n",
       " 'ENSG00000233755': 'lincRNA',\n",
       " 'ENSG00000233419': 'processed_pseudogene',\n",
       " 'ENSG00000231953': 'antisense',\n",
       " 'ENSG00000117614': 'processed_transcript',\n",
       " 'ENSG00000117616': 'processed_transcript',\n",
       " 'ENSG00000272432': 'lincRNA',\n",
       " 'ENSG00000261349': 'unprocessed_pseudogene',\n",
       " 'ENSG00000187010': 'processed_transcript',\n",
       " 'ENSG00000224183': 'processed_pseudogene',\n",
       " 'ENSG00000259984': 'unprocessed_pseudogene',\n",
       " 'ENSG00000183726': 'processed_transcript',\n",
       " 'ENSG00000252515': 'snRNA',\n",
       " 'ENSG00000238889': 'snoRNA',\n",
       " 'ENSG00000188672': 'processed_transcript',\n",
       " 'ENSG00000238084': 'processed_pseudogene',\n",
       " 'ENSG00000266468': 'miRNA',\n",
       " 'ENSG00000204178': 'processed_transcript',\n",
       " 'ENSG00000157978': 'processed_transcript',\n",
       " 'ENSG00000225643': 'lincRNA',\n",
       " 'ENSG00000117643': 'processed_transcript',\n",
       " 'ENSG00000233478': 'antisense',\n",
       " 'ENSG00000162430': 'processed_transcript',\n",
       " 'ENSG00000255054': 'nonsense_mediated_decay',\n",
       " 'ENSG00000228172': 'retained_intron',\n",
       " 'ENSG00000117640': 'retained_intron',\n",
       " 'ENSG00000223474': 'protein_coding',\n",
       " 'ENSG00000272478': 'antisense',\n",
       " 'ENSG00000127423': 'processed_transcript',\n",
       " 'ENSG00000236528': 'antisense',\n",
       " 'ENSG00000182749': 'protein_coding',\n",
       " 'ENSG00000207302': 'misc_RNA',\n",
       " 'ENSG00000117632': 'protein_coding',\n",
       " 'ENSG00000229786': 'processed_pseudogene',\n",
       " 'ENSG00000223624': 'processed_pseudogene',\n",
       " 'ENSG00000264021': 'miRNA',\n",
       " 'ENSG00000266763': 'miRNA',\n",
       " 'ENSG00000158006': 'protein_coding',\n",
       " 'ENSG00000207237': 'snRNA',\n",
       " 'ENSG00000252190': 'snoRNA',\n",
       " 'ENSG00000252691': 'snoRNA',\n",
       " 'ENSG00000158008': 'processed_transcript',\n",
       " 'ENSG00000158014': 'processed_transcript',\n",
       " 'ENSG00000158022': 'processed_transcript',\n",
       " 'ENSG00000265320': 'miRNA',\n",
       " 'ENSG00000175087': 'protein_coding',\n",
       " 'ENSG00000197245': 'protein_coding',\n",
       " 'ENSG00000142684': 'protein_coding',\n",
       " 'ENSG00000236782': 'antisense',\n",
       " 'ENSG00000142675': 'retained_intron',\n",
       " 'ENSG00000188782': 'nonsense_mediated_decay',\n",
       " 'ENSG00000236155': 'processed_transcript',\n",
       " 'ENSG00000130695': 'processed_transcript',\n",
       " 'ENSG00000270733': 'processed_pseudogene',\n",
       " 'ENSG00000142669': 'protein_coding',\n",
       " 'ENSG00000158062': 'protein_coding',\n",
       " 'ENSG00000169442': 'processed_transcript',\n",
       " 'ENSG00000176092': 'protein_coding',\n",
       " 'ENSG00000225854': 'processed_pseudogene',\n",
       " 'ENSG00000265143': 'misc_RNA',\n",
       " 'ENSG00000176083': 'protein_coding',\n",
       " 'ENSG00000131914': 'protein_coding',\n",
       " 'ENSG00000117682': 'protein_coding',\n",
       " 'ENSG00000223583': 'processed_pseudogene',\n",
       " 'ENSG00000225891': 'antisense',\n",
       " 'ENSG00000198830': 'processed_transcript',\n",
       " 'ENSG00000224129': 'processed_pseudogene',\n",
       " 'ENSG00000117676': 'protein_coding',\n",
       " 'ENSG00000238705': 'miRNA',\n",
       " 'ENSG00000243905': 'misc_RNA',\n",
       " 'ENSG00000252802': 'misc_RNA',\n",
       " 'ENSG00000229247': 'processed_pseudogene',\n",
       " 'ENSG00000238316': 'snoRNA',\n",
       " 'ENSG00000235069': 'processed_pseudogene',\n",
       " 'ENSG00000260063': 'antisense',\n",
       " 'ENSG00000117713': 'nonsense_mediated_decay',\n",
       " 'ENSG00000266138': 'misc_RNA',\n",
       " 'ENSG00000060642': 'protein_coding',\n",
       " 'ENSG00000241188': 'misc_RNA',\n",
       " 'ENSG00000204160': 'processed_transcript',\n",
       " 'ENSG00000175793': 'protein_coding',\n",
       " 'ENSG00000142751': 'processed_transcript',\n",
       " 'ENSG00000226698': 'antisense',\n",
       " 'ENSG00000198746': 'protein_coding',\n",
       " 'ENSG00000090273': 'processed_transcript',\n",
       " 'ENSG00000131910': 'protein_coding',\n",
       " 'ENSG00000175707': 'protein_coding',\n",
       " 'ENSG00000172746': 'unprocessed_pseudogene',\n",
       " 'ENSG00000224064': 'processed_pseudogene',\n",
       " 'ENSG00000225294': 'processed_pseudogene',\n",
       " 'ENSG00000253368': 'protein_coding',\n",
       " 'ENSG00000158246': 'protein_coding',\n",
       " 'ENSG00000090020': 'protein_coding',\n",
       " 'ENSG00000229985': 'processed_pseudogene',\n",
       " 'ENSG00000214973': 'processed_pseudogene',\n",
       " 'ENSG00000225159': 'processed_pseudogene',\n",
       " 'ENSG00000225990': 'processed_pseudogene',\n",
       " 'ENSG00000224311': 'antisense',\n",
       " 'ENSG00000142784': 'nonsense_mediated_decay',\n",
       " 'ENSG00000252098': 'misc_RNA',\n",
       " 'ENSG00000243659': 'processed_pseudogene',\n",
       " 'ENSG00000186501': 'retained_intron',\n",
       " 'ENSG00000206888': 'snRNA',\n",
       " 'ENSG00000241547': 'processed_pseudogene',\n",
       " 'ENSG00000142765': 'retained_intron',\n",
       " 'ENSG00000142733': 'retained_intron',\n",
       " 'ENSG00000142748': 'retained_intron',\n",
       " 'ENSG00000174950': 'protein_coding',\n",
       " 'ENSG00000231207': 'processed_pseudogene',\n",
       " 'ENSG00000181773': 'protein_coding',\n",
       " 'ENSG00000158195': 'protein_coding',\n",
       " 'ENSG00000241169': 'lincRNA',\n",
       " 'ENSG00000237429': 'antisense',\n",
       " 'ENSG00000126705': 'processed_transcript',\n",
       " 'ENSG00000000938': 'processed_transcript',\n",
       " 'ENSG00000235912': 'processed_pseudogene',\n",
       " 'ENSG00000233975': 'lincRNA',\n",
       " 'ENSG00000126709': 'protein_coding',\n",
       " 'ENSG00000225886': 'antisense',\n",
       " 'ENSG00000206767': 'snRNA',\n",
       " 'ENSG00000226662': 'processed_pseudogene',\n",
       " 'ENSG00000207095': 'snRNA',\n",
       " 'ENSG00000009780': 'protein_coding',\n",
       " 'ENSG00000231344': 'processed_pseudogene',\n",
       " 'ENSG00000230833': 'processed_pseudogene',\n",
       " 'ENSG00000117758': 'processed_transcript',\n",
       " 'ENSG00000269971': 'sense_intronic',\n",
       " 'ENSG00000270031': 'sense_intronic',\n",
       " 'ENSG00000223062': 'snRNA',\n",
       " 'ENSG00000117751': 'processed_transcript',\n",
       " 'ENSG00000252947': 'snoRNA',\n",
       " 'ENSG00000268552': 'protein_coding',\n",
       " 'ENSG00000130775': 'processed_transcript',\n",
       " 'ENSG00000117748': 'protein_coding',\n",
       " 'ENSG00000130768': 'protein_coding',\n",
       " 'ENSG00000227050': 'antisense',\n",
       " 'ENSG00000158156': 'processed_transcript',\n",
       " 'ENSG00000158161': 'processed_transcript',\n",
       " 'ENSG00000238231': 'processed_pseudogene',\n",
       " 'ENSG00000240750': 'misc_RNA',\n",
       " 'ENSG00000228589': 'processed_pseudogene',\n",
       " 'ENSG00000228943': 'processed_pseudogene',\n",
       " 'ENSG00000214812': 'processed_pseudogene',\n",
       " 'ENSG00000253005': 'snRNA',\n",
       " 'ENSG00000252429': 'snRNA',\n",
       " 'ENSG00000169403': 'protein_coding',\n",
       " 'ENSG00000126698': 'processed_transcript',\n",
       " 'ENSG00000215902': 'protein_coding',\n",
       " 'ENSG00000269804': 'protein_coding',\n",
       " 'ENSG00000130770': 'protein_coding',\n",
       " 'ENSG00000270605': 'antisense',\n",
       " 'ENSG00000271398': 'lincRNA',\n",
       " 'ENSG00000130766': 'protein_coding',\n",
       " 'ENSG00000130772': 'processed_transcript',\n",
       " 'ENSG00000204138': 'protein_coding',\n",
       " 'ENSG00000206779': 'misc_RNA',\n",
       " 'ENSG00000229820': 'processed_pseudogene',\n",
       " 'ENSG00000221216': 'snRNA',\n",
       " 'ENSG00000180198': 'retained_intron',\n",
       " 'ENSG00000242125': 'snoRNA',\n",
       " 'ENSG00000200087': 'snoRNA',\n",
       " 'ENSG00000238821': 'snoRNA',\n",
       " 'ENSG00000231468': 'processed_pseudogene',\n",
       " 'ENSG00000180098': 'processed_transcript',\n",
       " 'ENSG00000197989': 'snoRNA',\n",
       " 'ENSG00000221539': 'snoRNA',\n",
       " 'ENSG00000120656': 'processed_transcript',\n",
       " 'ENSG00000188060': 'processed_transcript',\n",
       " 'ENSG00000229388': 'lincRNA',\n",
       " 'ENSG00000270103': 'lincRNA',\n",
       " 'ENSG00000162419': 'processed_transcript',\n",
       " 'ENSG00000252777': 'snoRNA',\n",
       " 'ENSG00000198492': 'processed_transcript',\n",
       " 'ENSG00000116329': 'protein_coding',\n",
       " 'ENSG00000226852': 'lincRNA',\n",
       " 'ENSG00000233427': 'antisense',\n",
       " 'ENSG00000199756': 'misc_RNA',\n",
       " 'ENSG00000159023': 'processed_transcript',\n",
       " 'ENSG00000225616': 'processed_pseudogene',\n",
       " 'ENSG00000206704': 'misc_RNA',\n",
       " 'ENSG00000253304': 'protein_coding',\n",
       " 'ENSG00000225750': 'lincRNA',\n",
       " 'ENSG00000116350': 'processed_transcript',\n",
       " 'ENSG00000227637': 'antisense',\n",
       " 'ENSG00000116353': 'processed_transcript',\n",
       " 'ENSG00000237934': 'antisense',\n",
       " 'ENSG00000060656': 'retained_intron',\n",
       " 'ENSG00000221550': 'miRNA',\n",
       " 'ENSG00000230523': 'lincRNA',\n",
       " 'ENSG00000266678': 'miRNA',\n",
       " 'ENSG00000225011': 'unprocessed_pseudogene',\n",
       " 'ENSG00000221126': 'miRNA',\n",
       " 'ENSG00000228176': 'lincRNA',\n",
       " 'ENSG00000222787': 'miRNA',\n",
       " 'ENSG00000270927': 'processed_pseudogene',\n",
       " 'ENSG00000233399': 'lincRNA',\n",
       " 'ENSG00000233372': 'lincRNA',\n",
       " 'ENSG00000231251': 'processed_pseudogene',\n",
       " 'ENSG00000236335': 'lincRNA',\n",
       " 'ENSG00000231949': 'lincRNA',\n",
       " 'ENSG00000162510': 'processed_transcript',\n",
       " 'ENSG00000186056': 'antisense',\n",
       " 'ENSG00000162511': 'processed_transcript',\n",
       " 'ENSG00000264773': 'miRNA',\n",
       " 'ENSG00000229607': 'lincRNA',\n",
       " 'ENSG00000223382': 'lincRNA',\n",
       " 'ENSG00000222784': 'misc_RNA',\n",
       " 'ENSG00000235143': 'antisense',\n",
       " ...}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine enst_to_ensg and enst_to_gene to create dictionary relating gene_id and gene_name\n",
    "keys = list(enst_to_ensg.values())\n",
    "values = [enst_to_gene[enst] for enst in enst_to_ensg]\n",
    "ensg_to_gene = dict(zip(keys, values))\n",
    "ensg_to_gene\n",
    "\n",
    "# combine enst_to_ensg and enst_to_type to create dictionary relating gene_id and gene_name\n",
    "keys = list(enst_to_ensg.values())\n",
    "values = [ucsc_enst_to_type[enst] for enst in enst_to_ensg]\n",
    "ucsc_ensg_to_type = dict(zip(keys, values))\n",
    "ucsc_ensg_to_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gencode ensg_to_type length:  64854\n",
      "ucsc ensg_to_type length:  60234\n",
      "---\n",
      "number of shared ENSG keys:  55439\n",
      "differing values for same key:  24989 / 55439\n",
      "---\n",
      "non-null gene_names in kuehn_df after mapping gencode dict:  15704 / 15870\n",
      "non-null gene_names in kuehn_df after mapping ucsc dict:  15309 / 15870\n"
     ]
    }
   ],
   "source": [
    "# comparing and contrasting the dictionaries\n",
    "print('gencode ensg_to_type length: ', len(gencode_ensg_to_type))\n",
    "print('ucsc ensg_to_type length: ', len(ucsc_ensg_to_type))\n",
    "\n",
    "# checking the overlap of the two above dictionaries\n",
    "common_keys = set(gencode_ensg_to_type.keys()) & set(ucsc_ensg_to_type.keys()) # length 55439\n",
    "new_dict1 = {key: gencode_ensg_to_type[key] for key in common_keys}\n",
    "new_dict2 = {key: ucsc_ensg_to_type[key] for key in common_keys}\n",
    " \n",
    "# compare values of corresponding keys\n",
    "count = 0\n",
    "for key in new_dict1.keys():\n",
    "    if new_dict1[key] != new_dict2[key]:\n",
    "        # print(new_dict1[key], ' ', new_dict2[key])\n",
    "        count +=1\n",
    "\n",
    "print('---')\n",
    "print('number of shared ENSG keys: ', len(common_keys))\n",
    "print('differing values for same key: ', count, '/', len(common_keys))\n",
    "print('---')\n",
    "print('non-null gene_names in kuehn_df after mapping gencode dict: ', \n",
    "      kuehn_df['gene_ensemble'].map(gencode_ensg_to_type).notnull().sum(),\n",
    "      '/', len(kuehn_df))\n",
    "print('non-null gene_names in kuehn_df after mapping ucsc dict: ', \n",
    "      kuehn_df['gene_ensemble'].map(ucsc_ensg_to_type).notnull().sum(),\n",
    "      '/', len(kuehn_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ensemble</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>HC1_1</th>\n",
       "      <th>HC2_1</th>\n",
       "      <th>HC3_1</th>\n",
       "      <th>B.I.1_1</th>\n",
       "      <th>B.II.2_1</th>\n",
       "      <th>C.II.1_1</th>\n",
       "      <th>C.III.1_1</th>\n",
       "      <th>HC1_2</th>\n",
       "      <th>HC2_2</th>\n",
       "      <th>HC3_2</th>\n",
       "      <th>B.I.1_2</th>\n",
       "      <th>B.II.2_2</th>\n",
       "      <th>C.II.1_2</th>\n",
       "      <th>C.III.1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000225630</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>1134</td>\n",
       "      <td>1195</td>\n",
       "      <td>657</td>\n",
       "      <td>1223</td>\n",
       "      <td>671</td>\n",
       "      <td>2475</td>\n",
       "      <td>1992</td>\n",
       "      <td>1199</td>\n",
       "      <td>790</td>\n",
       "      <td>750</td>\n",
       "      <td>974</td>\n",
       "      <td>536</td>\n",
       "      <td>2189</td>\n",
       "      <td>794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>88</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>98</td>\n",
       "      <td>57</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>71</td>\n",
       "      <td>44</td>\n",
       "      <td>25</td>\n",
       "      <td>66</td>\n",
       "      <td>44</td>\n",
       "      <td>42</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000229344</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>25</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>25</td>\n",
       "      <td>29</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000248527</td>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>6605</td>\n",
       "      <td>6262</td>\n",
       "      <td>4856</td>\n",
       "      <td>6234</td>\n",
       "      <td>3118</td>\n",
       "      <td>5544</td>\n",
       "      <td>4551</td>\n",
       "      <td>8658</td>\n",
       "      <td>3438</td>\n",
       "      <td>4987</td>\n",
       "      <td>5283</td>\n",
       "      <td>3060</td>\n",
       "      <td>4876</td>\n",
       "      <td>1557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15865</th>\n",
       "      <td>ENSG00000276345</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>38</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15866</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15867</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>152</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>421</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>137</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15868</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>32</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15869</th>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>58</td>\n",
       "      <td>93</td>\n",
       "      <td>45</td>\n",
       "      <td>110</td>\n",
       "      <td>44</td>\n",
       "      <td>27</td>\n",
       "      <td>54</td>\n",
       "      <td>90</td>\n",
       "      <td>67</td>\n",
       "      <td>41</td>\n",
       "      <td>88</td>\n",
       "      <td>38</td>\n",
       "      <td>29</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15870 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_ensemble     gene_name               gene_type  HC1_1  HC2_1  \\\n",
       "0      ENSG00000279457           NaN                     NaN     17      6   \n",
       "1      ENSG00000225630      MTND2P28  unprocessed_pseudogene   1134   1195   \n",
       "2      ENSG00000237973  hsa-mir-6723  unprocessed_pseudogene     88     79   \n",
       "3      ENSG00000229344  RP5-857K21.7  unprocessed_pseudogene     25     20   \n",
       "4      ENSG00000248527      MTATP6P1  unprocessed_pseudogene   6605   6262   \n",
       "...                ...           ...                     ...    ...    ...   \n",
       "15865  ENSG00000276345           NaN                     NaN      0      0   \n",
       "15866  ENSG00000277856           NaN                     NaN      0      0   \n",
       "15867  ENSG00000275063           NaN                     NaN      1      2   \n",
       "15868  ENSG00000271254           NaN                     NaN      6     12   \n",
       "15869  ENSG00000275405           NaN                     NaN     58     93   \n",
       "\n",
       "       HC3_1  B.I.1_1  B.II.2_1  C.II.1_1  C.III.1_1  HC1_2  HC2_2  HC3_2  \\\n",
       "0          8        5        12         4          6     12     10      6   \n",
       "1        657     1223       671      2475       1992   1199    790    750   \n",
       "2         37       98        57        40         32     71     44     25   \n",
       "3         23       20         2        21          9     35     21     25   \n",
       "4       4856     6234      3118      5544       4551   8658   3438   4987   \n",
       "...      ...      ...       ...       ...        ...    ...    ...    ...   \n",
       "15865     64       54         0         0          2      0      1     69   \n",
       "15866      0       14        11         0          8      1      0      1   \n",
       "15867    152        7         2         5        421      0      0    137   \n",
       "15868     15       15         4         5         32      7      4     14   \n",
       "15869     45      110        44        27         54     90     67     41   \n",
       "\n",
       "       B.I.1_2  B.II.2_2  C.II.1_2  C.III.1_2  \n",
       "0           17         3        12          3  \n",
       "1          974       536      2189        794  \n",
       "2           66        44        42         28  \n",
       "3           29        18        15          5  \n",
       "4         5283      3060      4876       1557  \n",
       "...        ...       ...       ...        ...  \n",
       "15865       38         1         2          0  \n",
       "15866        6        11         0          6  \n",
       "15867        1         1         1        150  \n",
       "15868        6        10        11          3  \n",
       "15869       88        38        29         16  \n",
       "\n",
       "[15870 rows x 17 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kuehn_df.insert(1, 'gene_name', kuehn_df['gene_ensemble'].map(ensg_to_gene))\n",
    "\n",
    "# used gencode_ensg_to_type because more coverage of entries. also, \n",
    "#   if use ucsc_ensg_to_type, all missing gene_names will also have missing gene_types\n",
    "kuehn_df.insert(2, 'gene_type', kuehn_df['gene_ensemble'].map(gencode_ensg_to_type)) \n",
    "kuehn_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of missing genes:  561\n"
     ]
    }
   ],
   "source": [
    "missing_genes = kuehn_df['gene_type'][kuehn_df['gene_name'].isnull()]\n",
    "print('# of missing genes: ', len(missing_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_type\n",
       "lncRNA                                184\n",
       "protein_coding                        144\n",
       "TEC                                    86\n",
       "transcribed_unprocessed_pseudogene     19\n",
       "processed_pseudogene                   14\n",
       "unprocessed_pseudogene                  7\n",
       "artifact                                6\n",
       "transcribed_processed_pseudogene        5\n",
       "IG_V_gene                               4\n",
       "transcribed_unitary_pseudogene          1\n",
       "TR_C_gene                               1\n",
       "IG_V_pseudogene                         1\n",
       "snoRNA                                  1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the gene_types that I am able to identify for entries with missing gene_name for their\n",
    "#   given gene_ensemble ID\n",
    "missing_genes[missing_genes.notnull()].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ^ gene / transcript biotype descriptions for gencode genes found [here](https://www.gencodegenes.org/pages/biotypes.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Removing the genes we do not care about\n",
    "TODO: for now, just remove the nones. but come back to look at those later!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ensemble</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>HC1_1</th>\n",
       "      <th>HC2_1</th>\n",
       "      <th>HC3_1</th>\n",
       "      <th>B.I.1_1</th>\n",
       "      <th>B.II.2_1</th>\n",
       "      <th>C.II.1_1</th>\n",
       "      <th>C.III.1_1</th>\n",
       "      <th>HC1_2</th>\n",
       "      <th>HC2_2</th>\n",
       "      <th>HC3_2</th>\n",
       "      <th>B.I.1_2</th>\n",
       "      <th>B.II.2_2</th>\n",
       "      <th>C.II.1_2</th>\n",
       "      <th>C.III.1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000225630</td>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>1134</td>\n",
       "      <td>1195</td>\n",
       "      <td>657</td>\n",
       "      <td>1223</td>\n",
       "      <td>671</td>\n",
       "      <td>2475</td>\n",
       "      <td>1992</td>\n",
       "      <td>1199</td>\n",
       "      <td>790</td>\n",
       "      <td>750</td>\n",
       "      <td>974</td>\n",
       "      <td>536</td>\n",
       "      <td>2189</td>\n",
       "      <td>794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000237973</td>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>88</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>98</td>\n",
       "      <td>57</td>\n",
       "      <td>40</td>\n",
       "      <td>32</td>\n",
       "      <td>71</td>\n",
       "      <td>44</td>\n",
       "      <td>25</td>\n",
       "      <td>66</td>\n",
       "      <td>44</td>\n",
       "      <td>42</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000229344</td>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>25</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>35</td>\n",
       "      <td>21</td>\n",
       "      <td>25</td>\n",
       "      <td>29</td>\n",
       "      <td>18</td>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000248527</td>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>6605</td>\n",
       "      <td>6262</td>\n",
       "      <td>4856</td>\n",
       "      <td>6234</td>\n",
       "      <td>3118</td>\n",
       "      <td>5544</td>\n",
       "      <td>4551</td>\n",
       "      <td>8658</td>\n",
       "      <td>3438</td>\n",
       "      <td>4987</td>\n",
       "      <td>5283</td>\n",
       "      <td>3060</td>\n",
       "      <td>4876</td>\n",
       "      <td>1557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>RP5-857K21.11</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>34</td>\n",
       "      <td>21</td>\n",
       "      <td>10</td>\n",
       "      <td>22</td>\n",
       "      <td>19</td>\n",
       "      <td>16</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15852</th>\n",
       "      <td>ENSG00000198695</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1408</td>\n",
       "      <td>1838</td>\n",
       "      <td>626</td>\n",
       "      <td>1806</td>\n",
       "      <td>947</td>\n",
       "      <td>1132</td>\n",
       "      <td>893</td>\n",
       "      <td>1489</td>\n",
       "      <td>951</td>\n",
       "      <td>702</td>\n",
       "      <td>1676</td>\n",
       "      <td>754</td>\n",
       "      <td>1165</td>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15853</th>\n",
       "      <td>ENSG00000210194</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>121</td>\n",
       "      <td>36</td>\n",
       "      <td>17</td>\n",
       "      <td>21</td>\n",
       "      <td>82</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>36</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>30</td>\n",
       "      <td>13</td>\n",
       "      <td>5</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15854</th>\n",
       "      <td>ENSG00000198727</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>30727</td>\n",
       "      <td>38866</td>\n",
       "      <td>18227</td>\n",
       "      <td>37262</td>\n",
       "      <td>16825</td>\n",
       "      <td>26548</td>\n",
       "      <td>21676</td>\n",
       "      <td>39375</td>\n",
       "      <td>23277</td>\n",
       "      <td>18759</td>\n",
       "      <td>28383</td>\n",
       "      <td>14665</td>\n",
       "      <td>24045</td>\n",
       "      <td>8243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15855</th>\n",
       "      <td>ENSG00000210195</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>267</td>\n",
       "      <td>239</td>\n",
       "      <td>131</td>\n",
       "      <td>229</td>\n",
       "      <td>170</td>\n",
       "      <td>53</td>\n",
       "      <td>46</td>\n",
       "      <td>226</td>\n",
       "      <td>55</td>\n",
       "      <td>124</td>\n",
       "      <td>167</td>\n",
       "      <td>95</td>\n",
       "      <td>47</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15856</th>\n",
       "      <td>ENSG00000210196</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>146</td>\n",
       "      <td>93</td>\n",
       "      <td>47</td>\n",
       "      <td>98</td>\n",
       "      <td>106</td>\n",
       "      <td>53</td>\n",
       "      <td>62</td>\n",
       "      <td>63</td>\n",
       "      <td>32</td>\n",
       "      <td>30</td>\n",
       "      <td>110</td>\n",
       "      <td>34</td>\n",
       "      <td>51</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_ensemble      gene_name               gene_type  HC1_1  HC2_1  \\\n",
       "1      ENSG00000225630       MTND2P28  unprocessed_pseudogene   1134   1195   \n",
       "2      ENSG00000237973   hsa-mir-6723  unprocessed_pseudogene     88     79   \n",
       "3      ENSG00000229344   RP5-857K21.7  unprocessed_pseudogene     25     20   \n",
       "4      ENSG00000248527       MTATP6P1  unprocessed_pseudogene   6605   6262   \n",
       "5      ENSG00000198744  RP5-857K21.11  unprocessed_pseudogene     34     21   \n",
       "...                ...            ...                     ...    ...    ...   \n",
       "15852  ENSG00000198695         MT-ND6          protein_coding   1408   1838   \n",
       "15853  ENSG00000210194          MT-TE                 Mt_tRNA    121     36   \n",
       "15854  ENSG00000198727         MT-CYB          protein_coding  30727  38866   \n",
       "15855  ENSG00000210195          MT-TT                 Mt_tRNA    267    239   \n",
       "15856  ENSG00000210196          MT-TP                 Mt_tRNA    146     93   \n",
       "\n",
       "       HC3_1  B.I.1_1  B.II.2_1  C.II.1_1  C.III.1_1  HC1_2  HC2_2  HC3_2  \\\n",
       "1        657     1223       671      2475       1992   1199    790    750   \n",
       "2         37       98        57        40         32     71     44     25   \n",
       "3         23       20         2        21          9     35     21     25   \n",
       "4       4856     6234      3118      5544       4551   8658   3438   4987   \n",
       "5         10       22        19        16          5     22      3      5   \n",
       "...      ...      ...       ...       ...        ...    ...    ...    ...   \n",
       "15852    626     1806       947      1132        893   1489    951    702   \n",
       "15853     17       21        82         6          8     36      6     18   \n",
       "15854  18227    37262     16825     26548      21676  39375  23277  18759   \n",
       "15855    131      229       170        53         46    226     55    124   \n",
       "15856     47       98       106        53         62     63     32     30   \n",
       "\n",
       "       B.I.1_2  B.II.2_2  C.II.1_2  C.III.1_2  \n",
       "1          974       536      2189        794  \n",
       "2           66        44        42         28  \n",
       "3           29        18        15          5  \n",
       "4         5283      3060      4876       1557  \n",
       "5            9        18        13          5  \n",
       "...        ...       ...       ...        ...  \n",
       "15852     1676       754      1165        363  \n",
       "15853       30        13         5         34  \n",
       "15854    28383     14665     24045       8243  \n",
       "15855      167        95        47         57  \n",
       "15856      110        34        51         88  \n",
       "\n",
       "[15309 rows x 17 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kuehn_df = kuehn_df[kuehn_df['gene_name'].notnull()]\n",
    "kuehn_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### PCA on the data to see if we can get a separation between conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>MTND2P28</th>\n",
       "      <th>hsa-mir-6723</th>\n",
       "      <th>RP5-857K21.7</th>\n",
       "      <th>MTATP6P1</th>\n",
       "      <th>RP5-857K21.11</th>\n",
       "      <th>RP11-206L10.2</th>\n",
       "      <th>RP11-206L10.9</th>\n",
       "      <th>RP11-206L10.8</th>\n",
       "      <th>LINC01128</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>...</th>\n",
       "      <th>MT-ND4</th>\n",
       "      <th>MT-TH</th>\n",
       "      <th>MT-TS2</th>\n",
       "      <th>MT-TL2</th>\n",
       "      <th>MT-ND5</th>\n",
       "      <th>MT-ND6</th>\n",
       "      <th>MT-TE</th>\n",
       "      <th>MT-CYB</th>\n",
       "      <th>MT-TT</th>\n",
       "      <th>MT-TP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HC1_1</th>\n",
       "      <td>1134</td>\n",
       "      <td>88</td>\n",
       "      <td>25</td>\n",
       "      <td>6605</td>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>122</td>\n",
       "      <td>18</td>\n",
       "      <td>...</td>\n",
       "      <td>70603</td>\n",
       "      <td>382</td>\n",
       "      <td>159</td>\n",
       "      <td>70</td>\n",
       "      <td>65834</td>\n",
       "      <td>1408</td>\n",
       "      <td>121</td>\n",
       "      <td>30727</td>\n",
       "      <td>267</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2_1</th>\n",
       "      <td>1195</td>\n",
       "      <td>79</td>\n",
       "      <td>20</td>\n",
       "      <td>6262</td>\n",
       "      <td>21</td>\n",
       "      <td>70</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>170</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>88882</td>\n",
       "      <td>150</td>\n",
       "      <td>66</td>\n",
       "      <td>32</td>\n",
       "      <td>94145</td>\n",
       "      <td>1838</td>\n",
       "      <td>36</td>\n",
       "      <td>38866</td>\n",
       "      <td>239</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3_1</th>\n",
       "      <td>657</td>\n",
       "      <td>37</td>\n",
       "      <td>23</td>\n",
       "      <td>4856</td>\n",
       "      <td>10</td>\n",
       "      <td>26</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>90</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>57735</td>\n",
       "      <td>80</td>\n",
       "      <td>17</td>\n",
       "      <td>27</td>\n",
       "      <td>51021</td>\n",
       "      <td>626</td>\n",
       "      <td>17</td>\n",
       "      <td>18227</td>\n",
       "      <td>131</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.I.1_1</th>\n",
       "      <td>1223</td>\n",
       "      <td>98</td>\n",
       "      <td>20</td>\n",
       "      <td>6234</td>\n",
       "      <td>22</td>\n",
       "      <td>61</td>\n",
       "      <td>19</td>\n",
       "      <td>23</td>\n",
       "      <td>167</td>\n",
       "      <td>19</td>\n",
       "      <td>...</td>\n",
       "      <td>85596</td>\n",
       "      <td>74</td>\n",
       "      <td>43</td>\n",
       "      <td>27</td>\n",
       "      <td>81184</td>\n",
       "      <td>1806</td>\n",
       "      <td>21</td>\n",
       "      <td>37262</td>\n",
       "      <td>229</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.II.2_1</th>\n",
       "      <td>671</td>\n",
       "      <td>57</td>\n",
       "      <td>2</td>\n",
       "      <td>3118</td>\n",
       "      <td>19</td>\n",
       "      <td>35</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>80</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>45582</td>\n",
       "      <td>238</td>\n",
       "      <td>248</td>\n",
       "      <td>21</td>\n",
       "      <td>44593</td>\n",
       "      <td>947</td>\n",
       "      <td>82</td>\n",
       "      <td>16825</td>\n",
       "      <td>170</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  15309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name  MTND2P28  hsa-mir-6723  RP5-857K21.7  MTATP6P1  RP5-857K21.11  \\\n",
       "HC1_1          1134            88            25      6605             34   \n",
       "HC2_1          1195            79            20      6262             21   \n",
       "HC3_1           657            37            23      4856             10   \n",
       "B.I.1_1        1223            98            20      6234             22   \n",
       "B.II.2_1        671            57             2      3118             19   \n",
       "\n",
       "gene_name  RP11-206L10.2  RP11-206L10.9  RP11-206L10.8  LINC01128  LINC00115  \\\n",
       "HC1_1                 36             19             13        122         18   \n",
       "HC2_1                 70             10             16        170          1   \n",
       "HC3_1                 26             19              9         90          7   \n",
       "B.I.1_1               61             19             23        167         19   \n",
       "B.II.2_1              35             13             20         80         11   \n",
       "\n",
       "gene_name  ...  MT-ND4  MT-TH  MT-TS2  MT-TL2  MT-ND5  MT-ND6  MT-TE  MT-CYB  \\\n",
       "HC1_1      ...   70603    382     159      70   65834    1408    121   30727   \n",
       "HC2_1      ...   88882    150      66      32   94145    1838     36   38866   \n",
       "HC3_1      ...   57735     80      17      27   51021     626     17   18227   \n",
       "B.I.1_1    ...   85596     74      43      27   81184    1806     21   37262   \n",
       "B.II.2_1   ...   45582    238     248      21   44593     947     82   16825   \n",
       "\n",
       "gene_name  MT-TT  MT-TP  \n",
       "HC1_1        267    146  \n",
       "HC2_1        239     93  \n",
       "HC3_1        131     47  \n",
       "B.I.1_1      229     98  \n",
       "B.II.2_1     170    106  \n",
       "\n",
       "[5 rows x 15309 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose df to get samples in rows and genes (features) in columns\n",
    "numeric_T = kuehn_df.select_dtypes(include='number').T\n",
    "numeric_T.columns = kuehn_df['gene_name']\n",
    "numeric_T.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MTND2P28</th>\n",
       "      <th>hsa-mir-6723</th>\n",
       "      <th>RP5-857K21.7</th>\n",
       "      <th>MTATP6P1</th>\n",
       "      <th>RP5-857K21.11</th>\n",
       "      <th>RP11-206L10.2</th>\n",
       "      <th>RP11-206L10.9</th>\n",
       "      <th>RP11-206L10.8</th>\n",
       "      <th>LINC01128</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>...</th>\n",
       "      <th>MT-ND4</th>\n",
       "      <th>MT-TH</th>\n",
       "      <th>MT-TS2</th>\n",
       "      <th>MT-TL2</th>\n",
       "      <th>MT-ND5</th>\n",
       "      <th>MT-ND6</th>\n",
       "      <th>MT-TE</th>\n",
       "      <th>MT-CYB</th>\n",
       "      <th>MT-TT</th>\n",
       "      <th>MT-TP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>...</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>1.400000e+01</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>1.400000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>-1.268826e-16</td>\n",
       "      <td>-1.110223e-16</td>\n",
       "      <td>4.758099e-17</td>\n",
       "      <td>1.427430e-16</td>\n",
       "      <td>-9.516197e-17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.758099e-17</td>\n",
       "      <td>-6.344132e-17</td>\n",
       "      <td>-1.110223e-16</td>\n",
       "      <td>1.348128e-16</td>\n",
       "      <td>...</td>\n",
       "      <td>1.586033e-16</td>\n",
       "      <td>2.379049e-17</td>\n",
       "      <td>-4.758099e-17</td>\n",
       "      <td>1.189525e-17</td>\n",
       "      <td>-1.586033e-16</td>\n",
       "      <td>1.586033e-17</td>\n",
       "      <td>2.874685e-17</td>\n",
       "      <td>3.172066e-17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.149874e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749e+00</td>\n",
       "      <td>1.037749</td>\n",
       "      <td>1.037749e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>-1.103076e+00</td>\n",
       "      <td>-1.282119e+00</td>\n",
       "      <td>-1.974122e+00</td>\n",
       "      <td>-1.966150e+00</td>\n",
       "      <td>-1.335375e+00</td>\n",
       "      <td>-1.863381</td>\n",
       "      <td>-1.403161e+00</td>\n",
       "      <td>-1.663151e+00</td>\n",
       "      <td>-1.643365e+00</td>\n",
       "      <td>-1.658076e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.205759e+00</td>\n",
       "      <td>-1.001029e+00</td>\n",
       "      <td>-8.847461e-01</td>\n",
       "      <td>-1.315955e+00</td>\n",
       "      <td>-2.233670e+00</td>\n",
       "      <td>-1.718690e+00</td>\n",
       "      <td>-8.197891e-01</td>\n",
       "      <td>-1.824763e+00</td>\n",
       "      <td>-1.162372</td>\n",
       "      <td>-1.258332e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>-7.218919e-01</td>\n",
       "      <td>-7.114002e-01</td>\n",
       "      <td>-3.907117e-01</td>\n",
       "      <td>-7.077449e-01</td>\n",
       "      <td>-9.848391e-01</td>\n",
       "      <td>-0.492214</td>\n",
       "      <td>-7.068160e-01</td>\n",
       "      <td>-6.162033e-01</td>\n",
       "      <td>-4.328166e-01</td>\n",
       "      <td>-7.133582e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.725315e-01</td>\n",
       "      <td>-6.478331e-01</td>\n",
       "      <td>-6.437900e-01</td>\n",
       "      <td>-5.538372e-01</td>\n",
       "      <td>-5.582558e-01</td>\n",
       "      <td>-7.584115e-01</td>\n",
       "      <td>-6.854160e-01</td>\n",
       "      <td>-7.082075e-01</td>\n",
       "      <td>-1.039872</td>\n",
       "      <td>-7.235937e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>-2.215876e-01</td>\n",
       "      <td>-4.316361e-01</td>\n",
       "      <td>1.562847e-01</td>\n",
       "      <td>4.995406e-04</td>\n",
       "      <td>8.346094e-03</td>\n",
       "      <td>-0.175791</td>\n",
       "      <td>-1.570702e-01</td>\n",
       "      <td>-1.555465e-01</td>\n",
       "      <td>-2.656137e-01</td>\n",
       "      <td>-2.185061e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.410607e-01</td>\n",
       "      <td>-3.204178e-01</td>\n",
       "      <td>-3.637600e-01</td>\n",
       "      <td>-2.887527e-01</td>\n",
       "      <td>1.630107e-01</td>\n",
       "      <td>-1.883342e-01</td>\n",
       "      <td>-3.613396e-01</td>\n",
       "      <td>-1.231665e-01</td>\n",
       "      <td>-0.111448</td>\n",
       "      <td>-2.928326e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>5.579195e-02</td>\n",
       "      <td>7.209921e-01</td>\n",
       "      <td>6.169132e-01</td>\n",
       "      <td>6.590606e-01</td>\n",
       "      <td>7.094180e-01</td>\n",
       "      <td>0.439477</td>\n",
       "      <td>7.591727e-01</td>\n",
       "      <td>7.657672e-01</td>\n",
       "      <td>3.764453e-01</td>\n",
       "      <td>1.413863e-01</td>\n",
       "      <td>...</td>\n",
       "      <td>4.163801e-01</td>\n",
       "      <td>3.395692e-01</td>\n",
       "      <td>-7.396142e-02</td>\n",
       "      <td>1.420095e-01</td>\n",
       "      <td>4.446623e-01</td>\n",
       "      <td>7.753278e-01</td>\n",
       "      <td>1.445358e-01</td>\n",
       "      <td>5.919939e-01</td>\n",
       "      <td>0.978161</td>\n",
       "      <td>7.246547e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.196550e+00</td>\n",
       "      <td>1.985526e+00</td>\n",
       "      <td>1.826063e+00</td>\n",
       "      <td>2.172294e+00</td>\n",
       "      <td>2.286830e+00</td>\n",
       "      <td>2.074330</td>\n",
       "      <td>2.408410e+00</td>\n",
       "      <td>1.687081e+00</td>\n",
       "      <td>1.673940e+00</td>\n",
       "      <td>2.120795e+00</td>\n",
       "      <td>...</td>\n",
       "      <td>1.777813e+00</td>\n",
       "      <td>2.824832e+00</td>\n",
       "      <td>2.345368e+00</td>\n",
       "      <td>2.991668e+00</td>\n",
       "      <td>1.770200e+00</td>\n",
       "      <td>1.608171e+00</td>\n",
       "      <td>2.847807e+00</td>\n",
       "      <td>1.611097e+00</td>\n",
       "      <td>1.687374</td>\n",
       "      <td>2.187757e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows  15309 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           MTND2P28  hsa-mir-6723  RP5-857K21.7      MTATP6P1  RP5-857K21.11  \\\n",
       "count  1.400000e+01  1.400000e+01  1.400000e+01  1.400000e+01   1.400000e+01   \n",
       "mean  -1.268826e-16 -1.110223e-16  4.758099e-17  1.427430e-16  -9.516197e-17   \n",
       "std    1.037749e+00  1.037749e+00  1.037749e+00  1.037749e+00   1.037749e+00   \n",
       "min   -1.103076e+00 -1.282119e+00 -1.974122e+00 -1.966150e+00  -1.335375e+00   \n",
       "25%   -7.218919e-01 -7.114002e-01 -3.907117e-01 -7.077449e-01  -9.848391e-01   \n",
       "50%   -2.215876e-01 -4.316361e-01  1.562847e-01  4.995406e-04   8.346094e-03   \n",
       "75%    5.579195e-02  7.209921e-01  6.169132e-01  6.590606e-01   7.094180e-01   \n",
       "max    2.196550e+00  1.985526e+00  1.826063e+00  2.172294e+00   2.286830e+00   \n",
       "\n",
       "       RP11-206L10.2  RP11-206L10.9  RP11-206L10.8     LINC01128  \\\n",
       "count      14.000000   1.400000e+01   1.400000e+01  1.400000e+01   \n",
       "mean        0.000000   4.758099e-17  -6.344132e-17 -1.110223e-16   \n",
       "std         1.037749   1.037749e+00   1.037749e+00  1.037749e+00   \n",
       "min        -1.863381  -1.403161e+00  -1.663151e+00 -1.643365e+00   \n",
       "25%        -0.492214  -7.068160e-01  -6.162033e-01 -4.328166e-01   \n",
       "50%        -0.175791  -1.570702e-01  -1.555465e-01 -2.656137e-01   \n",
       "75%         0.439477   7.591727e-01   7.657672e-01  3.764453e-01   \n",
       "max         2.074330   2.408410e+00   1.687081e+00  1.673940e+00   \n",
       "\n",
       "          LINC00115  ...        MT-ND4         MT-TH        MT-TS2  \\\n",
       "count  1.400000e+01  ...  1.400000e+01  1.400000e+01  1.400000e+01   \n",
       "mean   1.348128e-16  ...  1.586033e-16  2.379049e-17 -4.758099e-17   \n",
       "std    1.037749e+00  ...  1.037749e+00  1.037749e+00  1.037749e+00   \n",
       "min   -1.658076e+00  ... -2.205759e+00 -1.001029e+00 -8.847461e-01   \n",
       "25%   -7.133582e-01  ... -3.725315e-01 -6.478331e-01 -6.437900e-01   \n",
       "50%   -2.185061e-01  ... -1.410607e-01 -3.204178e-01 -3.637600e-01   \n",
       "75%    1.413863e-01  ...  4.163801e-01  3.395692e-01 -7.396142e-02   \n",
       "max    2.120795e+00  ...  1.777813e+00  2.824832e+00  2.345368e+00   \n",
       "\n",
       "             MT-TL2        MT-ND5        MT-ND6         MT-TE        MT-CYB  \\\n",
       "count  1.400000e+01  1.400000e+01  1.400000e+01  1.400000e+01  1.400000e+01   \n",
       "mean   1.189525e-17 -1.586033e-16  1.586033e-17  2.874685e-17  3.172066e-17   \n",
       "std    1.037749e+00  1.037749e+00  1.037749e+00  1.037749e+00  1.037749e+00   \n",
       "min   -1.315955e+00 -2.233670e+00 -1.718690e+00 -8.197891e-01 -1.824763e+00   \n",
       "25%   -5.538372e-01 -5.582558e-01 -7.584115e-01 -6.854160e-01 -7.082075e-01   \n",
       "50%   -2.887527e-01  1.630107e-01 -1.883342e-01 -3.613396e-01 -1.231665e-01   \n",
       "75%    1.420095e-01  4.446623e-01  7.753278e-01  1.445358e-01  5.919939e-01   \n",
       "max    2.991668e+00  1.770200e+00  1.608171e+00  2.847807e+00  1.611097e+00   \n",
       "\n",
       "           MT-TT         MT-TP  \n",
       "count  14.000000  1.400000e+01  \n",
       "mean    0.000000 -1.149874e-16  \n",
       "std     1.037749  1.037749e+00  \n",
       "min    -1.162372 -1.258332e+00  \n",
       "25%    -1.039872 -7.235937e-01  \n",
       "50%    -0.111448 -2.928326e-01  \n",
       "75%     0.978161  7.246547e-01  \n",
       "max     1.687374  2.187757e+00  \n",
       "\n",
       "[8 rows x 15309 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# standardize features (scale to have mean = 0 and stddev = 1)\n",
    "scaler = preprocessing.StandardScaler()\n",
    "scaler.set_output(transform='pandas')\n",
    "df_scaled = scaler.fit_transform(numeric_T)\n",
    "df_scaled.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HC1_1</th>\n",
       "      <td>196.584308</td>\n",
       "      <td>-35.998346</td>\n",
       "      <td>-36.399623</td>\n",
       "      <td>-22.853499</td>\n",
       "      <td>0.201904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2_1</th>\n",
       "      <td>69.233474</td>\n",
       "      <td>20.063064</td>\n",
       "      <td>107.962979</td>\n",
       "      <td>-16.551061</td>\n",
       "      <td>-24.974623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3_1</th>\n",
       "      <td>-14.872202</td>\n",
       "      <td>-56.023478</td>\n",
       "      <td>5.929703</td>\n",
       "      <td>59.048541</td>\n",
       "      <td>-13.430065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.I.1_1</th>\n",
       "      <td>37.914550</td>\n",
       "      <td>100.213047</td>\n",
       "      <td>-12.344547</td>\n",
       "      <td>35.829483</td>\n",
       "      <td>3.286421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.II.2_1</th>\n",
       "      <td>-63.434972</td>\n",
       "      <td>18.853222</td>\n",
       "      <td>-26.961703</td>\n",
       "      <td>-20.337389</td>\n",
       "      <td>-16.064442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.II.1_1</th>\n",
       "      <td>-42.983227</td>\n",
       "      <td>-4.867297</td>\n",
       "      <td>8.159756</td>\n",
       "      <td>-4.636925</td>\n",
       "      <td>55.761989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.III.1_1</th>\n",
       "      <td>16.910673</td>\n",
       "      <td>34.483008</td>\n",
       "      <td>-32.619009</td>\n",
       "      <td>-29.087186</td>\n",
       "      <td>-28.649704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC1_2</th>\n",
       "      <td>182.738251</td>\n",
       "      <td>-33.500832</td>\n",
       "      <td>-16.672945</td>\n",
       "      <td>-3.245857</td>\n",
       "      <td>5.386621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2_2</th>\n",
       "      <td>-48.948330</td>\n",
       "      <td>-16.310101</td>\n",
       "      <td>41.625989</td>\n",
       "      <td>-25.395937</td>\n",
       "      <td>3.906713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3_2</th>\n",
       "      <td>-38.115597</td>\n",
       "      <td>-57.714235</td>\n",
       "      <td>2.934682</td>\n",
       "      <td>49.127188</td>\n",
       "      <td>-12.169633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.I.1_2</th>\n",
       "      <td>-12.456323</td>\n",
       "      <td>66.692503</td>\n",
       "      <td>-10.263062</td>\n",
       "      <td>27.998403</td>\n",
       "      <td>-2.081442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.II.2_2</th>\n",
       "      <td>-94.840557</td>\n",
       "      <td>4.343152</td>\n",
       "      <td>-15.459098</td>\n",
       "      <td>-12.104057</td>\n",
       "      <td>-8.991956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.II.1_2</th>\n",
       "      <td>-36.984050</td>\n",
       "      <td>-2.110914</td>\n",
       "      <td>5.586631</td>\n",
       "      <td>-9.782349</td>\n",
       "      <td>59.708611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.III.1_2</th>\n",
       "      <td>-150.746000</td>\n",
       "      <td>-38.122794</td>\n",
       "      <td>-21.479754</td>\n",
       "      <td>-28.009356</td>\n",
       "      <td>-21.890393</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  PC1         PC2         PC3        PC4        PC5\n",
       "HC1_1      196.584308  -35.998346  -36.399623 -22.853499   0.201904\n",
       "HC2_1       69.233474   20.063064  107.962979 -16.551061 -24.974623\n",
       "HC3_1      -14.872202  -56.023478    5.929703  59.048541 -13.430065\n",
       "B.I.1_1     37.914550  100.213047  -12.344547  35.829483   3.286421\n",
       "B.II.2_1   -63.434972   18.853222  -26.961703 -20.337389 -16.064442\n",
       "C.II.1_1   -42.983227   -4.867297    8.159756  -4.636925  55.761989\n",
       "C.III.1_1   16.910673   34.483008  -32.619009 -29.087186 -28.649704\n",
       "HC1_2      182.738251  -33.500832  -16.672945  -3.245857   5.386621\n",
       "HC2_2      -48.948330  -16.310101   41.625989 -25.395937   3.906713\n",
       "HC3_2      -38.115597  -57.714235    2.934682  49.127188 -12.169633\n",
       "B.I.1_2    -12.456323   66.692503  -10.263062  27.998403  -2.081442\n",
       "B.II.2_2   -94.840557    4.343152  -15.459098 -12.104057  -8.991956\n",
       "C.II.1_2   -36.984050   -2.110914    5.586631  -9.782349  59.708611\n",
       "C.III.1_2 -150.746000  -38.122794  -21.479754 -28.009356 -21.890393"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO plot better. add labels. see if can output as df\n",
    "n_comp = 5\n",
    "pca = PCA(n_components=n_comp)\n",
    "pca.set_output(transform='pandas')\n",
    "scores = pca.fit_transform(df_scaled)\n",
    "scores.columns = ['PC' + str(i + 1) for i in range(n_comp)]\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "      <th>PC4</th>\n",
       "      <th>PC5</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HC1_1</th>\n",
       "      <td>196.584308</td>\n",
       "      <td>-35.998346</td>\n",
       "      <td>-36.399623</td>\n",
       "      <td>-22.853499</td>\n",
       "      <td>0.201904</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2_1</th>\n",
       "      <td>69.233474</td>\n",
       "      <td>20.063064</td>\n",
       "      <td>107.962979</td>\n",
       "      <td>-16.551061</td>\n",
       "      <td>-24.974623</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3_1</th>\n",
       "      <td>-14.872202</td>\n",
       "      <td>-56.023478</td>\n",
       "      <td>5.929703</td>\n",
       "      <td>59.048541</td>\n",
       "      <td>-13.430065</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.I.1_1</th>\n",
       "      <td>37.914550</td>\n",
       "      <td>100.213047</td>\n",
       "      <td>-12.344547</td>\n",
       "      <td>35.829483</td>\n",
       "      <td>3.286421</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.II.2_1</th>\n",
       "      <td>-63.434972</td>\n",
       "      <td>18.853222</td>\n",
       "      <td>-26.961703</td>\n",
       "      <td>-20.337389</td>\n",
       "      <td>-16.064442</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.II.1_1</th>\n",
       "      <td>-42.983227</td>\n",
       "      <td>-4.867297</td>\n",
       "      <td>8.159756</td>\n",
       "      <td>-4.636925</td>\n",
       "      <td>55.761989</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.III.1_1</th>\n",
       "      <td>16.910673</td>\n",
       "      <td>34.483008</td>\n",
       "      <td>-32.619009</td>\n",
       "      <td>-29.087186</td>\n",
       "      <td>-28.649704</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC1_2</th>\n",
       "      <td>182.738251</td>\n",
       "      <td>-33.500832</td>\n",
       "      <td>-16.672945</td>\n",
       "      <td>-3.245857</td>\n",
       "      <td>5.386621</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC2_2</th>\n",
       "      <td>-48.948330</td>\n",
       "      <td>-16.310101</td>\n",
       "      <td>41.625989</td>\n",
       "      <td>-25.395937</td>\n",
       "      <td>3.906713</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HC3_2</th>\n",
       "      <td>-38.115597</td>\n",
       "      <td>-57.714235</td>\n",
       "      <td>2.934682</td>\n",
       "      <td>49.127188</td>\n",
       "      <td>-12.169633</td>\n",
       "      <td>HC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.I.1_2</th>\n",
       "      <td>-12.456323</td>\n",
       "      <td>66.692503</td>\n",
       "      <td>-10.263062</td>\n",
       "      <td>27.998403</td>\n",
       "      <td>-2.081442</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B.II.2_2</th>\n",
       "      <td>-94.840557</td>\n",
       "      <td>4.343152</td>\n",
       "      <td>-15.459098</td>\n",
       "      <td>-12.104057</td>\n",
       "      <td>-8.991956</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.II.1_2</th>\n",
       "      <td>-36.984050</td>\n",
       "      <td>-2.110914</td>\n",
       "      <td>5.586631</td>\n",
       "      <td>-9.782349</td>\n",
       "      <td>59.708611</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C.III.1_2</th>\n",
       "      <td>-150.746000</td>\n",
       "      <td>-38.122794</td>\n",
       "      <td>-21.479754</td>\n",
       "      <td>-28.009356</td>\n",
       "      <td>-21.890393</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  PC1         PC2         PC3        PC4        PC5 labels\n",
       "HC1_1      196.584308  -35.998346  -36.399623 -22.853499   0.201904     HC\n",
       "HC2_1       69.233474   20.063064  107.962979 -16.551061 -24.974623     HC\n",
       "HC3_1      -14.872202  -56.023478    5.929703  59.048541 -13.430065     HC\n",
       "B.I.1_1     37.914550  100.213047  -12.344547  35.829483   3.286421      B\n",
       "B.II.2_1   -63.434972   18.853222  -26.961703 -20.337389 -16.064442      B\n",
       "C.II.1_1   -42.983227   -4.867297    8.159756  -4.636925  55.761989      C\n",
       "C.III.1_1   16.910673   34.483008  -32.619009 -29.087186 -28.649704      C\n",
       "HC1_2      182.738251  -33.500832  -16.672945  -3.245857   5.386621     HC\n",
       "HC2_2      -48.948330  -16.310101   41.625989 -25.395937   3.906713     HC\n",
       "HC3_2      -38.115597  -57.714235    2.934682  49.127188 -12.169633     HC\n",
       "B.I.1_2    -12.456323   66.692503  -10.263062  27.998403  -2.081442      B\n",
       "B.II.2_2   -94.840557    4.343152  -15.459098 -12.104057  -8.991956      B\n",
       "C.II.1_2   -36.984050   -2.110914    5.586631  -9.782349  59.708611      C\n",
       "C.III.1_2 -150.746000  -38.122794  -21.479754 -28.009356 -21.890393      C"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cond = [''] * scores.shape[0]\n",
    "for idx, label in enumerate(scores.index):\n",
    "    if 'HC' in label:\n",
    "        cond[idx] = 'HC'\n",
    "    elif 'B' in label:\n",
    "        cond[idx] = 'B'\n",
    "    else:\n",
    "        cond[idx] = 'C'\n",
    "    \n",
    "scores['labels'] = cond\n",
    "scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'explained_variance_ratio'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/2b/tpns1_4s25dcgbbb9n1p9s0c0000gn/T/ipykernel_23443/972142288.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mscores\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexplained_variance_ratio\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/DAL/lib/python3.11/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6295\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6296\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6297\u001b[0m         ):\n\u001b[1;32m   6298\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6299\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'explained_variance_ratio'"
     ]
    }
   ],
   "source": [
    "# TODO: issue with outputting as dataframe i guess. i want scree\n",
    "scores.explained_variance_ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: xlabel='PC1', ylabel='PC2'>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAj4AAAGwCAYAAACpYG+ZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAABB40lEQVR4nO3deXhU9f3//ddkm0y2gRDIZCRAkChqRDBoBJSlatAvi9QFEWvhtuUnBaRUFOXGlkiVyCJyN1Es1i+uiFoFrRYLWkURRVZBtAgSFiExLGEm+2Q59x/8mDomgQCZTCbn+biuc13k8/nMOe9zmDAvPmcZi2EYhgAAAEwgJNAFAAAANBeCDwAAMA2CDwAAMA2CDwAAMA2CDwAAMA2CDwAAMA2CDwAAMI2wQBfQ0tTW1urQoUOKjY2VxWIJdDkAAKARDMNQcXGxnE6nQkIantch+PzMoUOHlJycHOgyAADAWThw4IA6duzYYD/B52diY2MlnThwcXFxAa4GAAA0htvtVnJysvdzvCEEn585eXorLi6O4AMAQJA53WUqXNwMAABMg+ADAABMg+ADAABMg2t8AABooWpra+XxeAJdRosQHh6u0NDQc14PwQcAgBbI4/EoLy9PtbW1gS6lxWjTpo0cDsc5PWeP4AMAQAtjGIby8/MVGhqq5OTkUz6QzwwMw1BZWZkKCwslSUlJSWe9LoIPAAAtTHV1tcrKyuR0OhUVFRXocloEm80mSSosLFSHDh3O+rSXuSMkAAAtUE1NjSQpIiIiwJW0LCdDYFVV1Vmvg+ADAEALxXdG+mqK48GpLgA4U4YhlR6WjFrJFi+F8b9yIFgQfADgTLjzpW/fljb8TaqqkC4aLmX8H6ltl0BXBqARWsyprk8++UTDhg2T0+mUxWLRihUrfPoNw1BWVpacTqdsNpsGDhyoHTt2+IyprKzUvffeq4SEBEVHR2v48OH64YcfmnEvALRqxfnSa3dJKx+UjuySXAekL56Snv2FdCwv0NUBXgMHDtSUKVMaNfbjjz+WxWLR8ePHz2mbXbp00cKFC89pHc2hxQSf0tJSXXbZZcrNza23f+7cuVqwYIFyc3O1YcMGORwOXX/99SouLvaOmTJlipYvX65ly5Zp7dq1Kikp0dChQ70XiQHAOTn0lXRwQ932sqPSulypurL5awJwRlrMqa4bb7xRN954Y719hmFo4cKFmjFjhm6++WZJ0gsvvKDExEQtXbpU99xzj1wul5577jm99NJLuu666yRJL7/8spKTk/XBBx9o8ODB9a67srJSlZX//cfK7XY38Z4BaBVqq6WtrzTc/81yqf9UKc7ZfDUBOGMtZsbnVPLy8lRQUKDMzExvm9Vq1YABA7Ru3TpJ0qZNm1RVVeUzxul0Ki0tzTumPtnZ2bLb7d4lOTnZfzsCIIiFSGHWhrvDrAqSf1JhMi+//LJ69+6t2NhYORwOjR492vsgwJ/67LPPdNlllykyMlIZGRnavn27T/+6devUv39/2Ww2JScna/LkySotLW1wu1lZWerUqZOsVqucTqcmT57c5Pt2NoLit7SgoECSlJiY6NOemJjo7SsoKFBERITatm3b4Jj6TJ8+XS6Xy7scOHCgiasH0CqEhEi97264v9evpej2zVcP0Egej0d//vOf9dVXX2nFihXKy8vT2LFj64x74IEHNH/+fG3YsEEdOnTQ8OHDvc/L2b59uwYPHqybb75Z27Zt02uvvaa1a9dq0qRJ9W7z73//u5588kn99a9/1a5du7RixQpdeuml/tzNRmsxp7oa4+f37xuGcdp7+k83xmq1ymo9xf/iAOCkhFTp0pHS9td/1n6BdPmvpSb4AkWgqd19938De9euXfWXv/xFV155pUpKShQTE+Ptmzlzpq6//npJJy4n6dixo5YvX66RI0dq3rx5Gj16tPeC6dTUVP3lL3/RgAEDtGjRIkVGRvpsc//+/XI4HLruuusUHh6uTp066corr/T/zjZCUMz4OBwOSaozc1NYWOidBXI4HPJ4PCoqKmpwDACck+j20uDZ0ph/SBcNk86/Vrr5WenXKyT7eYGuDqjXli1bdNNNN6lz586KjY3VwIEDJZ0IJz/Vp08f75/j4+N14YUX6ttvv5V04nKS559/XjExMd5l8ODBqq2tVV5e3Tsab7vtNpWXl6tr164aN26cli9frurqav/t5BkIiuCTkpIih8Oh1atXe9s8Ho/WrFmjvn37SpLS09MVHh7uMyY/P19ff/21dwwAnLOY9lJKf+mW56TbX5Z6jJTiCD1omUpLS5WZmamYmBi9/PLL2rBhg5YvXy7pxOfo6Zw8Y1JbW6t77rlHW7du9S5fffWVdu3apfPPP7/O65KTk7Vz50499dRTstlsmjBhgvr3739OXzXRVFrMqa6SkhLt3r3b+3NeXp62bt2q+Ph4derUSVOmTNHs2bOVmpqq1NRUzZ49W1FRURo9erQkyW636ze/+Y2mTp2qdu3aKT4+Xvfff78uvfRS711eANBkTnWhM9BC/Oc//9GRI0f0+OOPe2/e2bhxY71jv/jiC3Xq1EmSVFRUpO+++07du3eXJF1++eXasWOHunXr1uht22w2DR8+XMOHD9fEiRPVvXt3bd++XZdffvk57tW5aTHBZ+PGjRo0aJD35/vuu0+SNGbMGD3//POaNm2aysvLNWHCBBUVFSkjI0OrVq1SbGys9zVPPvmkwsLCNHLkSJWXl+vaa6/V888/f9bf4AoAQDDr1KmTIiIilJOTo/Hjx+vrr7/Wn//853rHzpo1S+3atVNiYqJmzJihhIQEjRgxQpL04IMP6qqrrtLEiRM1btw4RUdH69tvv9Xq1auVk5NTZ13PP/+8ampqlJGRoaioKL300kuy2Wzq3LmzP3e3UVrMqa6BAwfKMIw6y/PPPy/pxHRbVlaW8vPzVVFRoTVr1igtLc1nHZGRkcrJydHRo0dVVlamf/zjH9yeDgAwrfbt2+v555/XG2+8oYsvvliPP/645s+fX+/Yxx9/XL///e+Vnp6u/Px8vfPOO95vh+/Ro4fWrFmjXbt26ZprrlGvXr30xz/+UUlJSfWuq02bNnr22WfVr18/9ejRQx9++KH+8Y9/qF27dn7b18ayGIZhBLqIlsTtdstut8vlcikuLi7Q5QAATKiiokJ5eXlKSUmpc8eUmZ3quDT287vFzPgAAAD4G8EHAACYBsEHAACYBsEHAACYBsEHAACYBsEHAACYBsEHAACYBsEHAACYBsEHAACYBsEHAAA0ibFjx3q/3+unPv74Y1ksFh0/flySZBiGFi9erIyMDMXExKhNmzbq3bu3Fi5cqLKyMr/W2GK+pBQAADQtV5lHR0o8cldUKc4WroToCNmjIgJdlu666y699dZbevjhh5Wbm6v27dvrq6++0sKFC9WlS5d6w1NTIfgAANAKHTpergff3KZPdx3xtvVPTdDjt/SQs40tYHW9/vrreuWVV7RixQrddNNN3vYuXbpo+PDhcrvdft0+p7oAAGhlXGWeOqFHkj7ZdUQPvblNrjJPgCqTXnnlFV144YU+oecki8Uiu93u1+0z4wMAQCtzpMRTJ/Sc9MmuIzpS4vHbKa93331XMTExPm01NTXeP+/atUsXXnihX7bdGAQfAABaGXdF1Sn7i0/Tfy4GDRqkRYsW+bStX79ev/rVrySduLDZYrH4bfunQ/ABAKCViYsMP2V/7Gn6z0V0dLS6devm0/bDDz94/3zBBRfo22+/9dv2T4drfAAAaGUSYiLUPzWh3r7+qQlKiAncnV2jR4/Wd999p7fffrtOn2EYcrlcft0+wQcAgFbGHhWhx2/pUSf89E9N0JxbegT0lvaRI0fq9ttv1x133KHs7Gxt3LhR+/bt07vvvqvrrrtOH330kV+3z6kuAABaIWcbm3Lu6KUjJR4VV1QpNjJcCTGBf46PxWLR0qVLtXjxYv3v//6vHn30UYWFhSk1NVW//vWvNXjwYP9u3zAMw69bCDJut1t2u10ul0txcXGBLgcAYEIVFRXKy8tTSkqKIiMjA11Oi3Gq49LYz29OdQEAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMImuBTXV2thx9+WCkpKbLZbOratatmzZql2tpa7xjDMJSVlSWn0ymbzaaBAwdqx44dAawaAADzGDt2rCwWi3dp166dbrjhBm3bti3QpXkFTfCZM2eOnnnmGeXm5urbb7/V3LlzNW/ePOXk5HjHzJ07VwsWLFBubq42bNggh8Oh66+/XsXFxQGsHACAACkvko58J/2wUTqy68TPfnbDDTcoPz9f+fn5+vDDDxUWFqahQ4f6fbuNFTTfzv7555/rpptu0pAhQyRJXbp00auvvqqNGzdKOjHbs3DhQs2YMUM333yzJOmFF15QYmKili5dqnvuuSdgtQMA0OxcB6W3J0l7/v3ftvOvlYbnSPbz/LZZq9Uqh8MhSXI4HHrwwQfVv39/HT58WO3bt/fbdhsraGZ8rr76an344Yf67rvvJElfffWV1q5dq//5n/+RJOXl5amgoECZmZne11itVg0YMEDr1q1rcL2VlZVyu90+CwAAQa28qG7okaTvP5TeubdZZn4kqaSkRK+88oq6deumdu3aNcs2TydoZnwefPBBuVwude/eXaGhoaqpqdFjjz2mO+64Q5JUUFAgSUpMTPR5XWJiovbt29fgerOzs/XII4/4r3AAAJpb6eG6oeek7z880W9r65dNv/vuu4qJiTlRRmmpkpKS9O677yokpGXMtbSMKhrhtdde08svv6ylS5dq8+bNeuGFFzR//ny98MILPuMsFovPz4Zh1Gn7qenTp8vlcnmXAwcO+KV+AACaTcVpzl6crv8cDBo0SFu3btXWrVu1fv16ZWZm6sYbbzzlJERzCpoZnwceeEAPPfSQRo0aJUm69NJLtW/fPmVnZ2vMmDHe84kFBQVKSkryvq6wsLDOLNBPWa1WWa1W/xYPAEBziow7t/5zEB0drW7dunl/Tk9Pl91u17PPPqtHH33Ub9ttrKCZ8SkrK6szTRYaGuq9nT0lJUUOh0OrV6/29ns8Hq1Zs0Z9+/Zt1loBAAio6PYnLmSuz/nXnuhvJhaLRSEhISovL2+2bZ5K0Mz4DBs2TI899pg6deqkSy65RFu2bNGCBQt09913SzpxYKdMmaLZs2crNTVVqampmj17tqKiojR69OgAVw8AQDOytT1x99Y79564puekk3d1+en6HunETUMnr7stKipSbm6uSkpKNGzYML9t80wETfDJycnRH//4R02YMEGFhYVyOp2655579Kc//ck7Ztq0aSovL9eECRNUVFSkjIwMrVq1SrGxsQGsHACAALCfJ9363IkLmSvcJ05vRbf3a+iRpPfff997yUlsbKy6d++uN954QwMHDvTrdhvLYhiGEegiWhK32y273S6Xy6W4OP+dAwUAoCEVFRXKy8tTSkqKIiMjA11Oi3Gq49LYz++gucYHAADgXBF8AACAaRB8AACAaRB8AACAaRB8AACAaRB8AACAaRB8AACAaRB8AACAaRB8AACAaRB8AACAaRB8AABAkyooKNC9996rrl27ymq1Kjk5WcOGDdOHH354+hf7WdB8SSkAADgzrkqXjlUcU7GnWLERsYqPjJfdavfrNvfu3at+/fqpTZs2mjt3rnr06KGqqir961//0sSJE/Wf//zHr9s/HYIPAACtUEFpgWaum6l1h9Z52/o5+ymrb5Yc0Q6/bXfChAmyWCz68ssvFR0d7W2/5JJLdPfdd/ttu43FqS4AAFoZV6WrTuiRpM8OfaasdVlyVbr8st1jx47p/fff18SJE31Cz0lt2rTxy3bPBMEHAIBW5ljFsTqh56TPDn2mYxXH/LLd3bt3yzAMde/e3S/rbwoEHwAAWpliT/E59Z8twzAkSRaLxS/rbwoEHwAAWpnYiNhz6j9bqampslgs+vbbb/2y/qZA8AHQeMX50p410gezpC//Jh3bI1WVB7oqAD8THxmvfs5+9fb1c/ZTfGS8f7YbH6/BgwfrqaeeUmlpaZ3+48eP+2W7Z4LgA6BxXD9IL90svThcWvuE9M+pUm5vafcHhB+ghbFb7crqm1Un/Jy8q8uft7Q//fTTqqmp0ZVXXqk333xTu3bt0rfffqu//OUv6tOnj9+221jczg7g9KrKpY/nSIXf+LbX1khvjJUmbZTiUwJSGoD6OaIdmtN/TrM/xyclJUWbN2/WY489pqlTpyo/P1/t27dXenq6Fi1a5NdtNwbBB8DplR6Rti2rv6+2Wtq3juADtEB2q93vQac+SUlJys3NVW5ubrNv+3Q41QXg9GqrpBpPw/2lhc1XCwCcA4IPgNOLiJESLmi4v3P9F1ECQEtD8AFwejEdpBvn1t/X8UqpbZdmLQcAzhbBB0DjdLxCGvMPKfGSEz9HxEh97pVGvngiGAFAEODiZgCNY42RUvpLd70tVZVJIaFSdAcpLCLQlQGt1sknIeOEpjgeBB8AZyamfaArAFq90NBQSZLH45HNZgtwNS1HWVmZJCk8PPys10HwAQCghQkLC1NUVJQOHz6s8PBwhYSY+8oUwzBUVlamwsJCtWnTxhsMzwbBBwCAFsZisSgpKUl5eXnat29foMtpMdq0aSOHw3FO6yD4AADQAkVERCg1NVUezymeoWUi4eHh5zTTcxLBBwCAFiokJESRkZGBLqNVMfdJQwAAYCpBFXwOHjyoX/3qV2rXrp2ioqLUs2dPbdq0ydtvGIaysrLkdDpls9k0cOBA7dixI4AVAwCAliRogk9RUZH69eun8PBwrVy5Ut98842eeOIJtWnTxjtm7ty5WrBggXJzc7VhwwY5HA5df/31Ki4uDlzhAACgxbAYQfJ0pIceekifffaZPv3003r7DcOQ0+nUlClT9OCDD0qSKisrlZiYqDlz5uiee+6p93WVlZWqrKz0/ux2u5WcnCyXy6W4uLim3xEAANDk3G637Hb7aT+/g2bG55133lHv3r112223qUOHDurVq5eeffZZb39eXp4KCgqUmZnpbbNarRowYIDWrVvX4Hqzs7Nlt9u9S3Jysl/3AwAABE7QBJ89e/Zo0aJFSk1N1b/+9S+NHz9ekydP1osvvihJKigokCQlJib6vC4xMdHbV5/p06fL5XJ5lwMHDvhvJwAAQEAFze3stbW16t27t2bPni1J6tWrl3bs2KFFixbp17/+tXecxWLxeZ1hGHXafspqtcpqtfqnaAAA0KIEzYxPUlKSLr74Yp+2iy66SPv375ck75Mcfz67U1hYWGcWCAAAmFPQBJ9+/fpp586dPm3fffedOnfuLElKSUmRw+HQ6tWrvf0ej0dr1qxR3759m7VWAADQMgXNqa4//OEP6tu3r2bPnq2RI0fqyy+/1OLFi7V48WJJJ05xTZkyRbNnz1ZqaqpSU1M1e/ZsRUVFafTo0QGuHgAAtARBE3yuuOIKLV++XNOnT9esWbOUkpKihQsX6s477/SOmTZtmsrLyzVhwgQVFRUpIyNDq1atUmxsbAArBwAALUXQPMenuTT2OQAAAKDlaHXP8QEAADhXBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaBB8AAGAaQRt8srOzZbFYNGXKFG+bYRjKysqS0+mUzWbTwIEDtWPHjsAVCQAAWpSgDD4bNmzQ4sWL1aNHD5/2uXPnasGCBcrNzdWGDRvkcDh0/fXXq7i4OECVAgCAliTogk9JSYnuvPNOPfvss2rbtq233TAMLVy4UDNmzNDNN9+stLQ0vfDCCyorK9PSpUsDWDEAAGgpgi74TJw4UUOGDNF1113n056Xl6eCggJlZmZ626xWqwYMGKB169Y1uL7Kykq53W6fBQAAtE5hgS7gTCxbtkybN2/Whg0b6vQVFBRIkhITE33aExMTtW/fvgbXmZ2drUceeaRpCwUAAC1S0Mz4HDhwQL///e/18ssvKzIyssFxFovF52fDMOq0/dT06dPlcrm8y4EDB5qsZgAA0LIEzYzPpk2bVFhYqPT0dG9bTU2NPvnkE+Xm5mrnzp2STsz8JCUleccUFhbWmQX6KavVKqvV6r/CAQBAixE0Mz7XXnuttm/frq1bt3qX3r17684779TWrVvVtWtXORwOrV692vsaj8ejNWvWqG/fvgGsHAAAtBRBM+MTGxurtLQ0n7bo6Gi1a9fO2z5lyhTNnj1bqampSk1N1ezZsxUVFaXRo0cHomQAANDCBE3waYxp06apvLxcEyZMUFFRkTIyMrRq1SrFxsYGujQAANACWAzDMAJdREvidrtlt9vlcrkUFxcX6HIAAEAjNPbzu1XN+AAwL1elSyVVJQqxhCjeGi9rGDctAKiL4AMgqHlqPNp9fLfmfDlHmws3KzwkXEO7DtX4y8bLGeMMdHkAWpiguasLAOqz17VXd/7zTm0u3CxJqqqt0vLdy/XbVb9VQWlBgKsD0NIQfAAErRJPif6y5S+qrq2u03eg+IC2Hd4WgKoAtGQEHwBBq7SqVF8WfNlg/wf7PxD3bwD4KYIPgKAVYglRXETDd28kRCac8itrAJgPwQdA0Gpna6dfXfSrBvtHdBvRfMUACAoEHwBBK8QSoiFdhyjDkVGnb9oV05QUk1TPqwCYGbezAwhq7aPaa07/Ofqh+Aet+WGNYiNiNSB5gNrb2is2gqe2A/BF8AEQ9NrZ2qmdrZ0u63BZoEsB0MJxqgsAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJjGGQefr776So8++qiefvppHTlyxKfP7Xbr7rvvbrLiAAAAmpLFMAyjsYNXrVqlYcOGKTU1VcXFxSorK9Prr7+uQYMGSZJ+/PFHOZ1O1dTU+K1gf3O73bLb7XK5XIqLiwt0OQAAoBEa+/l9RjM+WVlZuv/++/X1119r7969mjZtmoYPH67333//nAsGAADwt7AzGbxjxw699NJLkiSLxaIHHnhAHTt21K233qpXX31VV155pV+KBAAAaApnFHysVquOHz/u03bHHXcoJCREo0aN0hNPPNGUtQEAADSpMwo+PXv21EcffaT09HSf9ttvv121tbUaM2ZMkxYHAADQlM4o+Pzud7/TJ598Um/fHXfcIUlavHjxuVcFAADgB2d0V5cZcFcXgGBXUlGlo6UeHTperqiIMHWIsyoxNlIhIZZAlwb4jV/u6ioqKlJOTo7cbnedPpfL1WBfU8jOztYVV1yh2NhYdejQQSNGjNDOnTt9xhiGoaysLDmdTtlsNg0cOFA7duzwSz0A0BIdKanUE6u/06D5H+uOZ9frpqc+0/Ccz7Tth+OqqeX/ucAZBZ/c3Fx98skn9SYpu92uTz/9VDk5OU1W3E+tWbNGEydO1BdffKHVq1erurpamZmZKi0t9Y6ZO3euFixYoNzcXG3YsEEOh0PXX3+9iouL/VITTKbaI5UVSVUVga4EqJdhGFq5PV9LPturn2acwyWVGv239Tp0vDxwxQEtxBmd6urZs6eeeOIJXXvttfX2f/jhh7r//vu1ZcuWJiuwIYcPH1aHDh20Zs0a9e/fX4ZhyOl0asqUKXrwwQclSZWVlUpMTNScOXN0zz33NGq9nOpCHVUVUtFeaf0zUv5XUkKq1GeSFN9VssYEujrAq8BdoeE5a1VYXFlv/7xbe+i23snNXBXQPPxyquv7779Xampqg/2pqan6/vvvz2SVZ83lckmS4uPjJUl5eXkqKChQZmamd4zVatWAAQO0bt26BtdTWVkpt9vtswBehiHt/0J6pq+0aYl0aLO07TVpcX9p17+kGk+gKwS8qmtqGww9kvTdj8x+A2cUfEJDQ3Xo0KEG+w8dOqSQEP9/76lhGLrvvvt09dVXKy0tTZJUUFAgSUpMTPQZm5iY6O2rT3Z2tux2u3dJTuZ/Q/iJ4nxpxT1S7c++hsUwpHcmS8UNv7eA5hYRGqLkeFuD/Zclt2m+YoAW6oxSSq9evbRixYoG+5cvX65evXqda02nNWnSJG3btk2vvvpqnT6LxfeuBcMw6rT91PTp0+VyubzLgQMHmrxeBLGyow2HG0+J5M5v3nqAU+gQF6kHMrvX2xcfHaFeyW2buSKg5Tmj4DNp0iQ98cQTys3N9fki0pqaGuXk5OjJJ5/UxIkTm7zIn7r33nv1zjvv6KOPPlLHjh297Q6HQ5LqzO4UFhbWmQX6KavVqri4OJ8F+K/TXQLHXTJoWa65IEF/HHqxoiNCvW3dHbF67f9cpfPaNjwbBJjFGT3A8JZbbtG0adM0efJkzZgxQ127dpXFYtH333+vkpISPfDAA7r11lv9UqhhGLr33nu1fPlyffzxx0pJSfHpT0lJkcPh0OrVq72zTh6PR2vWrNGcOXP8UhNMICpBiukglRTW7QuPkuKczV8TcAptoyJ011WdNPiSRB0vq1JEWIjaRUeoXYw10KUBLcJZPcBww4YNeuWVV7Rr1y4ZhqELLrhAo0eP9uuXlE6YMEFLly7V22+/rQsvvNDbbrfbZbOd+F/MnDlzlJ2drSVLlig1NVWzZ8/Wxx9/rJ07dyo2NrZR2+GuLviorZV2fyC9ertk1Pr2jXhaSrtVCuMDBQACrbGf32cUfMrKyvTAAw9oxYoVqqqq0rXXXqucnBwlJCQ0SdGn0tB1OkuWLNHYsWMlnZgVeuSRR/TXv/5VRUVFysjI0FNPPeW9ALoxCD6ow1MmHdsjfbZQKtgutesmXX2flHCBFNm4QA0A8C+/BJ8HHnhATz/9tO68807ZbDYtXbpUAwcO1BtvvNEkRbcEBB80qKpM8pSeOMUVER3oagAAP9HYz+8zusbnrbfe0nPPPadRo0ZJku68807169dPNTU1Cg0NPc2rgSAXHnViAQAErTO6q+vAgQO65pprvD9feeWVCgsLO+WzfQAAAFqKMwo+NTU1ioiI8GkLCwtTdXV1kxYFAADgD2d0qsswDI0dO1ZW63/vYqmoqND48eMVHf3fax7eeuutpqsQAACgiZxR8BkzZkydtl/96ldNVgwAAIA/nVHwWbJkib/qAAAA8Dv/f6MoAABAC0HwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAApkHwAQAAphEW6AIASZKnVKpwSyGhUkyHQFcDAGilCD4IrOoqqWiPtGautOcjydZGumqi1H2IFOsIdHUAgFaG4IPAOrJT+tu1UnXFiZ/Ljkrv3SftXCmNWCTFtA9sfQCAVoVrfBA45S7pXzP+G3p+avdq6fi+5q8JANCqEXwQOJUuKe/jhvv/816zlQIAMAeCDwLHEiKFRjTcHxHdfLUAAEyB4IPAiWonpd3WcH/3oc1XCwDAFAg+CJxwmzTwQcmeXLdv4P8rxSY1f00AgFaNu7oQWG07S3e/L+V9Kn2zQopuL/X+jRTfRbLZA10dAKCVIfgg8OwdpZ53SGm3nHiAYUhooCsCALRSBB+0HGGnuNAZAIAm0Cqv8Xn66aeVkpKiyMhIpaen69NPPw10SQAAoAVodcHntdde05QpUzRjxgxt2bJF11xzjW688Ubt378/0KUBAIAAsxiGYQS6iKaUkZGhyy+/XIsWLfK2XXTRRRoxYoSys7PrjK+srFRlZaX3Z7fbreTkZLlcLsXFxTVLzQAA4Ny43W7Z7fbTfn63qhkfj8ejTZs2KTMz06c9MzNT69atq/c12dnZstvt3iU5uZ5bq4EWpLCsULuLdivPlaeiiqJAlwMAQaVVXdx85MgR1dTUKDEx0ac9MTFRBQUF9b5m+vTpuu+++7w/n5zxAVqasqoybf5xs2Z9MUv5pfmSpIvbXaxH+z2qbm26yWKxBLhCAGj5WtWMz0k//wAwDKPBDwWr1aq4uDifBWiJvnd9rwkfTvCGHkn65ug3GrNyjA6WHAxgZQAQPFpV8ElISFBoaGid2Z3CwsI6s0BAMHFXupWzOUeG6l6SV1xVrI/2fxSAqgAg+LSq4BMREaH09HStXr3ap3316tXq27dvgKoCzl15dbl2HN3RYP/6gvWqqq1qxooAIDi1qmt8JOm+++7TXXfdpd69e6tPnz5avHix9u/fr/Hjxwe6NOCshYeEyxHtkNvjrre/S1wXhVla3a8zADS5Vvcv5e23366jR49q1qxZys/PV1pamv75z3+qc+fOgS4NOK2a2hodLj+sYxXHVF1brQRbghJsCYq3xeueHvdo6pqpdV5jkUW/TP0lFzcDQCO0uuf4nKvGPgcAaGqeGo+2FG7RtE+m6VjFMUmSLcym+3vfrxu63KDq2mr979f/qxe/edF7rY811Krsq7N1dcerZQuzBbJ8AAioxn5+E3x+huCDQNnr2qtfvvNLVddW1+l7LvM5XZl0pUo8JTpacVTfHftOkWGR6mrvqoSoBFlDrQGoGABajsZ+fre6U11AMKo1avX27rfrDT2S9NTWp3Rh/IWyW+2KiYhR5zhO3QLA2WhVd3UBwcpT49G3x75tsH+fe58qqiuasSIAaJ0IPkALEBEaoUsSLmmwP8WewjU8ANAECD5ACxBiCdHw84crPCS83v6JPScqzso1ZwBwrgg+QAvhjHbq2cxn1d7W3tsWHR6tP/f7sy6MvzCAlQFA68HFzUALER4arss7XK5Xh7yqY5XHVFtbq/jIeLW3tVdYKL+qANAU+NcUaEEsFosSoxOVGM13ywGAP3CqCwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmAbBBwAAmEZYoAsAzMYwDB0urlRNraHI8FC1jY4IdEkAYBoEH6AZHSmu0Ps7ftSij7/Xj+4KXdrRruk3dtfFTrtirPw6AoC/caoLaCbHyzx6fOVOPbziax08Xq7qWkNb9h/XyL9+ofV7jga6PAAwBYIP0EwOF1fq75t/qLfvT2/vUIG7opkrAgDzIfgAzeTrQ+4G+w4eL1dxeVUzVgMA5kTwAZpJjDX0lP1hoZZmqgQAzIvgAzST7o44WcPq/5Xr0zVe8dzdBQB+R/ABmklinFW5oy9XaIjvzE77GKtm33yp7DaCDwD4G/fPAs0kIixU16Qm6IP7+uu9bfnKO1Kqa1ITdEWXdjqvrS3Q5QGAKQTFjM/evXv1m9/8RikpKbLZbDr//PM1c+ZMeTwen3H79+/XsGHDFB0drYSEBE2ePLnOGCCQIsNDlZIQo0m/SNUTI3tqRK+OhB4AaEZBMePzn//8R7W1tfrrX/+qbt266euvv9a4ceNUWlqq+fPnS5Jqamo0ZMgQtW/fXmvXrtXRo0c1ZswYGYahnJycAO8BAABoCSyGYRiBLuJszJs3T4sWLdKePXskSStXrtTQoUN14MABOZ1OSdKyZcs0duxYFRYWKi4urlHrdbvdstvtcrlcjX4NAAAIrMZ+fgfFqa76uFwuxcfHe3/+/PPPlZaW5g09kjR48GBVVlZq06ZNDa6nsrJSbrfbZwEAAK1TUAaf77//Xjk5ORo/fry3raCgQImJiT7j2rZtq4iICBUUFDS4ruzsbNntdu+SnJzst7oBAEBgBTT4ZGVlyWKxnHLZuHGjz2sOHTqkG264Qbfddpt++9vf+vRZLHUfAGcYRr3tJ02fPl0ul8u7HDhwoGl2DgAAtDgBvbh50qRJGjVq1CnHdOnSxfvnQ4cOadCgQerTp48WL17sM87hcGj9+vU+bUVFRaqqqqozE/RTVqtVVqv1zIsHAABBJ6DBJyEhQQkJCY0ae/DgQQ0aNEjp6elasmSJQkJ8J6v69Omjxx57TPn5+UpKSpIkrVq1SlarVenp6U1eOwAACD5BcVfXoUOHNGDAAHXq1EkvvviiQkP/+51HDodD0onb2Xv27KnExETNmzdPx44d09ixYzVixIgzup2du7oAAAg+jf38Dorn+KxatUq7d+/W7t271bFjR5++k7ktNDRU7733niZMmKB+/frJZrNp9OjR3uf8AAAABMWMT3NixgcAgODT6p/jAwAAcKYIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDQIPgAAwDTCAl0AAABomTzVNTpc7NH+Y2WqqKpRSkK0EmIiFBMZHujSzhrBBwAA1FFZVaPPvj+qSUs3q8xTI0kKsUj39O+qcf3PV3x0RIArPDuc6gIAAHUcPF6ucS9u9IYeSao1pEVr9ujz748EsLJzQ/ABAAB1vLX5oGpqjXr7/vLhbh0tqWzmipoGwQcAAPioqa3VrsLiBvsPHi+Xp6a2GStqOgQfAADgIzQkRFd1bddg/0VJsbKFh57ROksqqnXgWJn2HS3VsVLPuZZ41ri4GQAA1HHdRYlasOo7FVdW1+l78IbuahPV+Iub9x4p1WP//EYffluoWkPq0dGuP9+UpouSYhURdmYB6lwx4wMAAOo4r41Nr4/vo+6OWG9b+xirFt15uU/b6RwsKtNtf/1cq785EXokadsPLt36zDrlHSlt6rJPixkfAABQR0iIRRclxenl32boeJlHVTWG2kaFq0NspEJCLI1ez5rvjuhwcd0LoatqDC38YJfm33aZoq3NF0cIPgAAoEEJMVYlxFjP6rWe6lqt/ubHBvvX5x1TcUU1wac1Kasq0+Hyw/r0h09VVFmkfs5+6hTbSQlRCYEuDQAAvwoNschhj2ywPz46QmFnMHvUFAg+flTqKdWqfas0c91MGTpxYnPxtsXq1b6X5g2Yp8ToxABXCACA/4SGWPSrqzrp1S/319s/vn9XJcSe3WzS2eLiZj8qLC/Un9b9yRt6TtpyeIte3/m6qmvrXikPAEBrktw2SjOHXSzLzyZ2hl2WpAEXtm/2epjx8aPVe1c32Lds5zKN7D5SiVHM+gAAWq84W7hu652sgRd20LrdR1ReVaN+3RKUGBcZkO/7Ivj40ZGKhr/LxO1xyzDqfxQ4AACtSYw1TDHWMKUkRAe6lOA71VVZWamePXvKYrFo69atPn379+/XsGHDFB0drYSEBE2ePFkeT+CeDtn/vP4N9qUnpssWZmvGagAAQNAFn2nTpsnpdNZpr6mp0ZAhQ1RaWqq1a9dq2bJlevPNNzV16tQAVHnCBfEX6Hz7+XXaQywhmtp7quxWewCqAgDAvIIq+KxcuVKrVq3S/Pnz6/StWrVK33zzjV5++WX16tVL1113nZ544gk9++yzcrvdDa6zsrJSbrfbZ2kqHaI6aNH1izTygpGyhp64aj0tIU0v3vCiUtukNtl2AABA4wTNNT4//vijxo0bpxUrVigqKqpO/+eff660tDSf2aDBgwersrJSmzZt0qBBg+pdb3Z2th555BG/1Z0UnaRpV0zTb3v8VrW1tYoKj1LbyLZ+2x4AAGhYUMz4GIahsWPHavz48erdu3e9YwoKCpSY6HuHVNu2bRUREaGCgoIG1z19+nS5XC7vcuDAgSatXZKsYVYlRSfpvNjzCD0AAARQQINPVlaWLBbLKZeNGzcqJydHbrdb06dPP+X6LD9/SIBOhKb62k+yWq2Ki4vzWQAAQOsU0FNdkyZN0qhRo045pkuXLnr00Uf1xRdfyGr1fbpj7969deedd+qFF16Qw+HQ+vXrffqLiopUVVVVZyYIAACYk8UIgofJ7N+/3+ei40OHDmnw4MH6+9//royMDHXs2FErV67U0KFD9cMPPygpKUmS9Nprr2nMmDEqLCxs9EyO2+2W3W6Xy+Vi9gcAgCDR2M/voLi4uVOnTj4/x8TESJLOP/98dezYUZKUmZmpiy++WHfddZfmzZunY8eO6f7779e4ceMIMAAAQFKQXNzcGKGhoXrvvfcUGRmpfv36aeTIkRoxYkS9t74DAABzCopTXc2JU10AAASfxn5+t5oZHwAAgNMh+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMg+AAAANMIC3QBAFq2QneF9h8r03c/Fqtj2yh16xCjJHukLBZLoEsDgDNG8AHQoAPHyjR2yQZ9f7jE2xYfHaFXfpuh7o5Ywg+AoMOpLgD1cpVX6f9dvt0n9EjSsVKPxi75UgXuigBVBgBnj+ADBJHjZR796K5QcUWV37d1rKRSn+46Um/fj+5K5bsIPgCCD6e6gCBwvMyjbT+49P99uEsHi8p1sTNOf7guVee3j1GU1T+/xuVVtafsLyr1+GW7AOBPBB+ghSurrNarX+7XnPd3etsK3BX6aGeh/vbr3vpF9w5+udYmzhamqIhQlXlq6u3vFB/V5NsEAH/jVBfQwh0p8eiJVd/VaTcMafpb2/Wju9Iv2+0Qa9WEgefX23fdRR3UPtbql+0CgD8x4wO0cHuPlqq61qi3r7C4UkVlHjnskU2+3YiwUN1xZSdFhocq59+75SqvkjUsRKOuTNaEgd3UJiqiybcJAP5G8AFauPDQU0/MhvrxlvJ2MVb9P3276H8uTVKZp0aR4SFqH2uVNSzUb9sEAH8KqlNd7733njIyMmSz2ZSQkKCbb77Zp3///v0aNmyYoqOjlZCQoMmTJ8vj4QJMBLfkeJts4fUHja4J0WoTHe7X7YeGhsjZxqZuHWLUsW0UoQdAUAuaGZ8333xT48aN0+zZs/WLX/xChmFo+/bt3v6amhoNGTJE7du319q1a3X06FGNGTNGhmEoJycngJUD56ZDrFULRl6mCUs3y/jJGa/I8BA9eXtPdYht+tNcANBaWQzDqP/igRakurpaXbp00SOPPKLf/OY39Y5ZuXKlhg4dqgMHDsjpdEqSli1bprFjx6qwsFBxcXGN2pbb7ZbdbpfL5Wr0awB/K/NU64dj5Xpl/T7tPlyi3p3jdfPl5+m8NjaFneZUGACYQWM/v4Nixmfz5s06ePCgQkJC1KtXLxUUFKhnz56aP3++LrnkEknS559/rrS0NG/okaTBgwersrJSmzZt0qBBg+pdd2VlpSor/3tXjNvt9u/OAGchKiJMFzhi9cehF8tTU6vIsFCFhPB1EQBwpoLiv4p79uyRJGVlZenhhx/Wu+++q7Zt22rAgAE6duyYJKmgoECJiYk+r2vbtq0iIiJUUFDQ4Lqzs7Nlt9u9S3Jysv92BDhHYaEhiooII/QAwFkKaPDJysqSxWI55bJx40bV1p54guyMGTN0yy23KD09XUuWLJHFYtEbb7zhXV99D3EzDOOUD3ebPn26XC6Xdzlw4EDT7ygAAGgRAnqqa9KkSRo1atQpx3Tp0kXFxcWSpIsvvtjbbrVa1bVrV+3fv1+S5HA4tH79ep/XFhUVqaqqqs5M0E9ZrVZZrTyIDQAAMwho8ElISFBCQsJpx6Wnp8tqtWrnzp26+uqrJUlVVVXau3evOnfuLEnq06ePHnvsMeXn5yspKUmStGrVKlmtVqWnp/tvJwAAQNAIioub4+LiNH78eM2cOVPJycnq3Lmz5s2bJ0m67bbbJEmZmZm6+OKLddddd2nevHk6duyY7r//fo0bN467swAAgKQgCT6SNG/ePIWFhemuu+5SeXm5MjIy9O9//1tt27aVJIWGhuq9997ThAkT1K9fP9lsNo0ePVrz588PcOUAAKClCIrn+DQnnuMDAEDwaeznd1Dczg4AANAUCD4AAMA0CD4AAMA0CD4AAMA0guauruZy8lpvvrMLAIDgcfJz+3T3bBF8fubkU6L5zi4AAIJPcXGx7HZ7g/3czv4ztbW1OnTokGJjY0/5HV9nw+12Kzk5WQcOHDDtrfIcA46BxDGQOAYSx0DiGEhNdwwMw1BxcbGcTqdCQhq+kocZn58JCQlRx44d/bqNuLg4077BT+IYcAwkjoHEMZA4BhLHQGqaY3CqmZ6TuLgZAACYBsEHAACYBsGnGVmtVs2cOVNWqzXQpQQMx4BjIHEMJI6BxDGQOAZS8x8DLm4GAACmwYwPAAAwDYIPAAAwDYIPAAAwDYIPAAAwDYKPnzz22GPq27evoqKi1KZNm3rHWCyWOsszzzzjM2b79u0aMGCAbDabzjvvPM2aNeu030PSEjRm//fv369hw4YpOjpaCQkJmjx5sjwej8+YYN3/hnTp0qXO3/lDDz3kM6YxxyWYPf3000pJSVFkZKTS09P16aefBrokv8nKyqrz9+1wOLz9hmEoKytLTqdTNptNAwcO1I4dOwJY8bn75JNPNGzYMDmdTlksFq1YscKnvzH7XFlZqXvvvVcJCQmKjo7W8OHD9cMPPzTjXpyb0x2DsWPH1nlfXHXVVT5jgvkYZGdn64orrlBsbKw6dOigESNGaOfOnT5jAvk+IPj4icfj0W233abf/e53pxy3ZMkS5efne5cxY8Z4+9xut66//no5nU5t2LBBOTk5mj9/vhYsWODv8s/Z6fa/pqZGQ4YMUWlpqdauXatly5bpzTff1NSpU71jgnn/T2XWrFk+f+cPP/ywt68xxyWYvfbaa5oyZYpmzJihLVu26JprrtGNN96o/fv3B7o0v7nkkkt8/r63b9/u7Zs7d64WLFig3NxcbdiwQQ6HQ9dff733OwODUWlpqS677DLl5ubW29+YfZ4yZYqWL1+uZcuWae3atSopKdHQoUNVU1PTXLtxTk53DCTphhtu8Hlf/POf//TpD+ZjsGbNGk2cOFFffPGFVq9ererqamVmZqq0tNQ7JqDvAwN+tWTJEsNut9fbJ8lYvnx5g699+umnDbvdblRUVHjbsrOzDafTadTW1jZxpf7R0P7/85//NEJCQoyDBw9621599VXDarUaLpfLMIzWsf8/17lzZ+PJJ59ssL8xxyWYXXnllcb48eN92rp372489NBDAarIv2bOnGlcdtll9fbV1tYaDofDePzxx71tFRUVht1uN5555plmqtC/fv5vXGP2+fjx40Z4eLixbNky75iDBw8aISEhxvvvv99stTeV+v6dHzNmjHHTTTc1+JrWdgwKCwsNScaaNWsMwwj8+4AZnwCbNGmSEhISdMUVV+iZZ55RbW2tt+/zzz/XgAEDfB7qNHjwYB06dEh79+4NQLVN5/PPP1daWpqcTqe3bfDgwaqsrNSmTZu8Y1rj/s+ZM0ft2rVTz5499dhjj/mcxmrMcQlWHo9HmzZtUmZmpk97Zmam1q1bF6Cq/G/Xrl1yOp1KSUnRqFGjtGfPHklSXl6eCgoKfI6H1WrVgAEDWu3xaMw+b9q0SVVVVT5jnE6n0tLSWtVx+fjjj9WhQwddcMEFGjdunAoLC719re0YuFwuSVJ8fLykwL8P+JLSAPrzn/+sa6+9VjabTR9++KGmTp2qI0eOeE99FBQUqEuXLj6vSUxM9PalpKQ0d8lNpqCgwLsvJ7Vt21YREREqKCjwjmlt+//73/9el19+udq2basvv/xS06dPV15env72t79JatxxCVZHjhxRTU1Nnf1LTEwM+n1rSEZGhl588UVdcMEF+vHHH/Xoo4+qb9++2rFjh3ef6zse+/btC0S5fteYfS4oKFBERITatm1bZ0xreZ/ceOONuu2229S5c2fl5eXpj3/8o37xi19o06ZNslqtreoYGIah++67T1dffbXS0tIkBf59wIzPGajvQsWfLxs3bmz0+h5++GH16dNHPXv21NSpUzVr1izNmzfPZ4zFYvH52fi/F/b+vL05NPX+17cPhmH4tLek/W/ImRyXP/zhDxowYIB69Oih3/72t3rmmWf03HPP6ejRo971Nea4BLP6/k5by7793I033qhbbrlFl156qa677jq99957kqQXXnjBO8ZMx+Oks9nn1nRcbr/9dg0ZMkRpaWkaNmyYVq5cqe+++877/mhIMB6DSZMmadu2bXr11Vfr9AXqfcCMzxmYNGmSRo0adcoxP5+hOBNXXXWV3G63fvzxRyUmJsrhcNRJtienQ3+elJtDU+6/w+HQ+vXrfdqKiopUVVXl3beWtv8NOZfjcvJOjt27d6tdu3aNOi7BKiEhQaGhofX+nQb7vjVWdHS0Lr30Uu3atUsjRoyQdOJ/tklJSd4xrfl4nLyj7VT77HA45PF4VFRU5PO//cLCQvXt27d5C24mSUlJ6ty5s3bt2iWp9RyDe++9V++8844++eQTdezY0dse6PcBMz5nICEhQd27dz/lEhkZedbr37JliyIjI723f/fp00effPKJzzUgq1atktPpPKeAdbaacv/79Omjr7/+Wvn5+d62VatWyWq1Kj093TumJe1/Q87luGzZskWSvL/8jTkuwSoiIkLp6elavXq1T/vq1auD6h/zc1FZWalvv/1WSUlJSklJkcPh8DkeHo9Ha9asabXHozH7nJ6ervDwcJ8x+fn5+vrrr1vtcTl69KgOHDjg/Xcg2I+BYRiaNGmS3nrrLf373/+uc1lCwN8H53RpNBq0b98+Y8uWLcYjjzxixMTEGFu2bDG2bNliFBcXG4ZhGO+8846xePFiY/v27cbu3buNZ5991oiLizMmT57sXcfx48eNxMRE44477jC2b99uvPXWW0ZcXJwxf/78QO1Wo51u/6urq420tDTj2muvNTZv3mx88MEHRseOHY1JkyZ51xHM+1+fdevWGQsWLDC2bNli7Nmzx3jttdcMp9NpDB8+3DumMcclmC1btswIDw83nnvuOeObb74xpkyZYkRHRxt79+4NdGl+MXXqVOPjjz829uzZY3zxxRfG0KFDjdjYWO/+Pv7444bdbjfeeustY/v27cYdd9xhJCUlGW63O8CVn73i4mLv77sk73t+3759hmE0bp/Hjx9vdOzY0fjggw+MzZs3G7/4xS+Myy67zKiurg7Ubp2RUx2D4uJiY+rUqca6deuMvLw846OPPjL69OljnHfeea3mGPzud78z7Ha78fHHHxv5+fnepayszDsmkO8Dgo+fjBkzxpBUZ/noo48MwzCMlStXGj179jRiYmKMqKgoIy0tzVi4cKFRVVXls55t27YZ11xzjWG1Wg2Hw2FkZWUFxa3cp9t/wzgRjoYMGWLYbDYjPj7emDRpks+t64YRvPtfn02bNhkZGRmG3W43IiMjjQsvvNCYOXOmUVpa6jOuMcclmD311FNG586djYiICOPyyy/33uLaGt1+++1GUlKSER4ebjidTuPmm282duzY4e2vra01Zs6caTgcDsNqtRr9+/c3tm/fHsCKz91HH31U7+/+mDFjDMNo3D6Xl5cbkyZNMuLj4w2bzWYMHTrU2L9/fwD25uyc6hiUlZUZmZmZRvv27Y3w8HCjU6dOxpgxY+rsXzAfg/r2XZKxZMkS75hAvg8s/7dIAACAVo9rfAAAgGkQfAAAgGkQfAAAgGkQfAAAgGkQfAAAgGkQfAAAgGkQfAAAgGkQfAAAgGkQfAAAgGkQfAAErbFjx8pischisSg8PFxdu3bV/fffr9LSUu+YN998UwMHDpTdbldMTIx69OihWbNm6dixY5JOfPHh6NGjdeGFFyokJERTpkwJ0N4AaA4EHwBB7YYbblB+fr727NmjRx99VE8//bTuv/9+SdKMGTN0++2364orrtDKlSv19ddf64knntBXX32ll156SdKJb0xv3769ZsyYocsuuyyQuwKgGfBdXQCC1tixY3X8+HGtWLHC2zZu3Di9++67evvtt5WRkaGFCxfq97//fZ3XHj9+XG3atPFpGzhwoHr27KmFCxf6t3AAAcOMD4BWxWazqaqqSq+88opiYmI0YcKEesf9PPQAMAeCD4BW48svv9TSpUt17bXXateuXeratavCw8MDXRaAFoTgAyCovfvuu4qJiVFkZKT69Omj/v37KycnR4ZhyGKxBLo8AC1MWKALAIBzMWjQIC1atEjh4eFyOp3eGZ4LLrhAa9euVVVVFbM+ALyY8QEQ1KKjo9WtWzd17tzZJ+CMHj1aJSUlevrpp+t93fHjx5upQgAtCTM+AFqljIwMTZs2TVOnTtXBgwf1y1/+Uk6nU7t379Yzzzyjq6++2nu319atWyVJJSUlOnz4sLZu3aqIiAhdfPHFAdwDAP7A7ewAglZ9t7P/3Ouvv66nnnpKW7ZsUW1trc4//3zdeuutuvfee713dtV3LVDnzp21d+9e/xQOIGAIPgAAwDS4xgcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJgGwQcAAJjG/w92oTmudoc13AAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.scatterplot(data=scores, x='PC1', y='PC2', hue='labels')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Plotting differential gene expression heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HC1_1        1658.370762\n",
       "HC2_1        1266.899275\n",
       "HC3_1         931.033967\n",
       "B.I.1_1      1158.363642\n",
       "B.II.2_1      764.612189\n",
       "C.II.1_1      807.492325\n",
       "C.III.1_1    1028.972500\n",
       "HC1_2        1589.922137\n",
       "HC2_2         776.485858\n",
       "HC3_2         843.779803\n",
       "B.I.1_2       964.438762\n",
       "B.II.2_2      631.168398\n",
       "C.II.1_2      829.769613\n",
       "C.III.1_2     413.008034\n",
       "dtype: float64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kuehn_df.mean(numeric_only=True, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HC1_1        7490.257376\n",
       "HC2_1        5641.916230\n",
       "HC3_1        4106.656490\n",
       "B.I.1_1      5398.647187\n",
       "B.II.2_1     4037.636068\n",
       "C.II.1_1     3412.541319\n",
       "C.III.1_1    4083.271228\n",
       "HC1_2        6467.719811\n",
       "HC2_2        3437.683081\n",
       "HC3_2        3711.930550\n",
       "B.I.1_2      4642.084018\n",
       "B.II.2_2     2772.768124\n",
       "C.II.1_2     3460.255412\n",
       "C.III.1_2    1958.613773\n",
       "dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ddof = 0 for population stddev, ddof = 1 for sample stddev\n",
    "kuehn_df.std(numeric_only=True, axis=0, ddof=1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>HC1_1</th>\n",
       "      <th>HC2_1</th>\n",
       "      <th>HC3_1</th>\n",
       "      <th>B.I.1_1</th>\n",
       "      <th>B.II.2_1</th>\n",
       "      <th>C.II.1_1</th>\n",
       "      <th>C.III.1_1</th>\n",
       "      <th>HC1_2</th>\n",
       "      <th>HC2_2</th>\n",
       "      <th>HC3_2</th>\n",
       "      <th>B.I.1_2</th>\n",
       "      <th>B.II.2_2</th>\n",
       "      <th>C.II.1_2</th>\n",
       "      <th>C.III.1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>-0.070007</td>\n",
       "      <td>-0.012744</td>\n",
       "      <td>-0.066729</td>\n",
       "      <td>0.011973</td>\n",
       "      <td>-0.023185</td>\n",
       "      <td>0.488641</td>\n",
       "      <td>0.235847</td>\n",
       "      <td>-0.060442</td>\n",
       "      <td>0.003931</td>\n",
       "      <td>-0.025264</td>\n",
       "      <td>0.002060</td>\n",
       "      <td>-0.034323</td>\n",
       "      <td>0.392812</td>\n",
       "      <td>0.194521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>-0.209655</td>\n",
       "      <td>-0.210549</td>\n",
       "      <td>-0.217704</td>\n",
       "      <td>-0.196413</td>\n",
       "      <td>-0.175254</td>\n",
       "      <td>-0.224903</td>\n",
       "      <td>-0.244160</td>\n",
       "      <td>-0.234847</td>\n",
       "      <td>-0.213075</td>\n",
       "      <td>-0.220581</td>\n",
       "      <td>-0.193542</td>\n",
       "      <td>-0.211763</td>\n",
       "      <td>-0.227662</td>\n",
       "      <td>-0.196572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>-0.218066</td>\n",
       "      <td>-0.221006</td>\n",
       "      <td>-0.221113</td>\n",
       "      <td>-0.210861</td>\n",
       "      <td>-0.188876</td>\n",
       "      <td>-0.230471</td>\n",
       "      <td>-0.249793</td>\n",
       "      <td>-0.240413</td>\n",
       "      <td>-0.219766</td>\n",
       "      <td>-0.220581</td>\n",
       "      <td>-0.201513</td>\n",
       "      <td>-0.221139</td>\n",
       "      <td>-0.235465</td>\n",
       "      <td>-0.208315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>0.660408</td>\n",
       "      <td>0.885355</td>\n",
       "      <td>0.955757</td>\n",
       "      <td>0.940168</td>\n",
       "      <td>0.582863</td>\n",
       "      <td>1.387971</td>\n",
       "      <td>0.862550</td>\n",
       "      <td>1.092824</td>\n",
       "      <td>0.774217</td>\n",
       "      <td>1.116190</td>\n",
       "      <td>0.930307</td>\n",
       "      <td>0.875959</td>\n",
       "      <td>1.169344</td>\n",
       "      <td>0.584082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RP5-857K21.11</td>\n",
       "      <td>-0.216864</td>\n",
       "      <td>-0.220829</td>\n",
       "      <td>-0.224278</td>\n",
       "      <td>-0.210490</td>\n",
       "      <td>-0.184666</td>\n",
       "      <td>-0.231936</td>\n",
       "      <td>-0.250773</td>\n",
       "      <td>-0.242423</td>\n",
       "      <td>-0.225002</td>\n",
       "      <td>-0.225969</td>\n",
       "      <td>-0.205821</td>\n",
       "      <td>-0.221139</td>\n",
       "      <td>-0.236043</td>\n",
       "      <td>-0.208315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15852</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>-0.033426</td>\n",
       "      <td>0.101225</td>\n",
       "      <td>-0.074278</td>\n",
       "      <td>0.119963</td>\n",
       "      <td>0.045172</td>\n",
       "      <td>0.095093</td>\n",
       "      <td>-0.033300</td>\n",
       "      <td>-0.015604</td>\n",
       "      <td>0.050765</td>\n",
       "      <td>-0.038196</td>\n",
       "      <td>0.153285</td>\n",
       "      <td>0.044299</td>\n",
       "      <td>0.096880</td>\n",
       "      <td>-0.025532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15853</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>-0.205249</td>\n",
       "      <td>-0.218170</td>\n",
       "      <td>-0.222574</td>\n",
       "      <td>-0.210676</td>\n",
       "      <td>-0.169062</td>\n",
       "      <td>-0.234867</td>\n",
       "      <td>-0.250038</td>\n",
       "      <td>-0.240258</td>\n",
       "      <td>-0.224129</td>\n",
       "      <td>-0.222466</td>\n",
       "      <td>-0.201297</td>\n",
       "      <td>-0.222943</td>\n",
       "      <td>-0.238355</td>\n",
       "      <td>-0.193508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15854</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>3.880859</td>\n",
       "      <td>6.664243</td>\n",
       "      <td>4.211690</td>\n",
       "      <td>6.687534</td>\n",
       "      <td>3.977671</td>\n",
       "      <td>7.542915</td>\n",
       "      <td>5.056492</td>\n",
       "      <td>5.842102</td>\n",
       "      <td>6.545256</td>\n",
       "      <td>4.826389</td>\n",
       "      <td>5.906520</td>\n",
       "      <td>5.061307</td>\n",
       "      <td>6.709109</td>\n",
       "      <td>3.997721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15855</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>-0.185757</td>\n",
       "      <td>-0.182190</td>\n",
       "      <td>-0.194814</td>\n",
       "      <td>-0.172148</td>\n",
       "      <td>-0.147267</td>\n",
       "      <td>-0.221094</td>\n",
       "      <td>-0.240732</td>\n",
       "      <td>-0.210881</td>\n",
       "      <td>-0.209876</td>\n",
       "      <td>-0.193910</td>\n",
       "      <td>-0.171785</td>\n",
       "      <td>-0.193369</td>\n",
       "      <td>-0.226217</td>\n",
       "      <td>-0.181765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15856</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>-0.201912</td>\n",
       "      <td>-0.208067</td>\n",
       "      <td>-0.215269</td>\n",
       "      <td>-0.196413</td>\n",
       "      <td>-0.163118</td>\n",
       "      <td>-0.221094</td>\n",
       "      <td>-0.236813</td>\n",
       "      <td>-0.236084</td>\n",
       "      <td>-0.216566</td>\n",
       "      <td>-0.219234</td>\n",
       "      <td>-0.184064</td>\n",
       "      <td>-0.215369</td>\n",
       "      <td>-0.225061</td>\n",
       "      <td>-0.165938</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name     HC1_1     HC2_1     HC3_1   B.I.1_1  B.II.2_1  \\\n",
       "1           MTND2P28 -0.070007 -0.012744 -0.066729  0.011973 -0.023185   \n",
       "2       hsa-mir-6723 -0.209655 -0.210549 -0.217704 -0.196413 -0.175254   \n",
       "3       RP5-857K21.7 -0.218066 -0.221006 -0.221113 -0.210861 -0.188876   \n",
       "4           MTATP6P1  0.660408  0.885355  0.955757  0.940168  0.582863   \n",
       "5      RP5-857K21.11 -0.216864 -0.220829 -0.224278 -0.210490 -0.184666   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "15852         MT-ND6 -0.033426  0.101225 -0.074278  0.119963  0.045172   \n",
       "15853          MT-TE -0.205249 -0.218170 -0.222574 -0.210676 -0.169062   \n",
       "15854         MT-CYB  3.880859  6.664243  4.211690  6.687534  3.977671   \n",
       "15855          MT-TT -0.185757 -0.182190 -0.194814 -0.172148 -0.147267   \n",
       "15856          MT-TP -0.201912 -0.208067 -0.215269 -0.196413 -0.163118   \n",
       "\n",
       "       C.II.1_1  C.III.1_1     HC1_2     HC2_2     HC3_2   B.I.1_2  B.II.2_2  \\\n",
       "1      0.488641   0.235847 -0.060442  0.003931 -0.025264  0.002060 -0.034323   \n",
       "2     -0.224903  -0.244160 -0.234847 -0.213075 -0.220581 -0.193542 -0.211763   \n",
       "3     -0.230471  -0.249793 -0.240413 -0.219766 -0.220581 -0.201513 -0.221139   \n",
       "4      1.387971   0.862550  1.092824  0.774217  1.116190  0.930307  0.875959   \n",
       "5     -0.231936  -0.250773 -0.242423 -0.225002 -0.225969 -0.205821 -0.221139   \n",
       "...         ...        ...       ...       ...       ...       ...       ...   \n",
       "15852  0.095093  -0.033300 -0.015604  0.050765 -0.038196  0.153285  0.044299   \n",
       "15853 -0.234867  -0.250038 -0.240258 -0.224129 -0.222466 -0.201297 -0.222943   \n",
       "15854  7.542915   5.056492  5.842102  6.545256  4.826389  5.906520  5.061307   \n",
       "15855 -0.221094  -0.240732 -0.210881 -0.209876 -0.193910 -0.171785 -0.193369   \n",
       "15856 -0.221094  -0.236813 -0.236084 -0.216566 -0.219234 -0.184064 -0.215369   \n",
       "\n",
       "       C.II.1_2  C.III.1_2  \n",
       "1      0.392812   0.194521  \n",
       "2     -0.227662  -0.196572  \n",
       "3     -0.235465  -0.208315  \n",
       "4      1.169344   0.584082  \n",
       "5     -0.236043  -0.208315  \n",
       "...         ...        ...  \n",
       "15852  0.096880  -0.025532  \n",
       "15853 -0.238355  -0.193508  \n",
       "15854  6.709109   3.997721  \n",
       "15855 -0.226217  -0.181765  \n",
       "15856 -0.225061  -0.165938  \n",
       "\n",
       "[15309 rows x 15 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this is what we are hoping to get information from, since combining datasets\n",
    "# zscore across genes per sample with ddof = 1 default for sp.stats.zscore\n",
    "zscored_df = kuehn_df.loc[:, 'HC1_1':].apply(sp.stats.zscore, axis=0, ddof=1)\n",
    "zscored_df.insert(0, 'gene_name', kuehn_df['gene_name'])\n",
    "zscored_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zscored across genes per sample\n",
    "# sns.set_context('paper', font_scale = 0.5) \n",
    "# sns.clustermap(zscored_df.loc[:, 'HC1_1':], xticklabels=zscored_df.columns[1:], yticklabels=zscored_df['gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>HC1_1</th>\n",
       "      <th>HC2_1</th>\n",
       "      <th>HC3_1</th>\n",
       "      <th>B.I.1_1</th>\n",
       "      <th>B.II.2_1</th>\n",
       "      <th>C.II.1_1</th>\n",
       "      <th>C.III.1_1</th>\n",
       "      <th>HC1_2</th>\n",
       "      <th>HC2_2</th>\n",
       "      <th>HC3_2</th>\n",
       "      <th>B.I.1_2</th>\n",
       "      <th>B.II.2_2</th>\n",
       "      <th>C.II.1_2</th>\n",
       "      <th>C.III.1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MTND2P28</td>\n",
       "      <td>-0.082342</td>\n",
       "      <td>0.017687</td>\n",
       "      <td>-0.864533</td>\n",
       "      <td>0.063601</td>\n",
       "      <td>-0.841576</td>\n",
       "      <td>2.116648</td>\n",
       "      <td>1.324618</td>\n",
       "      <td>0.024246</td>\n",
       "      <td>-0.646438</td>\n",
       "      <td>-0.712031</td>\n",
       "      <td>-0.344712</td>\n",
       "      <td>-1.062951</td>\n",
       "      <td>1.647661</td>\n",
       "      <td>-0.639879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-mir-6723</td>\n",
       "      <td>1.481961</td>\n",
       "      <td>1.093755</td>\n",
       "      <td>-0.717873</td>\n",
       "      <td>1.913301</td>\n",
       "      <td>0.144807</td>\n",
       "      <td>-0.588471</td>\n",
       "      <td>-0.933543</td>\n",
       "      <td>0.748683</td>\n",
       "      <td>-0.415935</td>\n",
       "      <td>-1.235481</td>\n",
       "      <td>0.533013</td>\n",
       "      <td>-0.415935</td>\n",
       "      <td>-0.502203</td>\n",
       "      <td>-1.106079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RP5-857K21.7</td>\n",
       "      <td>0.649957</td>\n",
       "      <td>0.095116</td>\n",
       "      <td>0.428020</td>\n",
       "      <td>0.095116</td>\n",
       "      <td>-1.902312</td>\n",
       "      <td>0.206084</td>\n",
       "      <td>-1.125534</td>\n",
       "      <td>1.759638</td>\n",
       "      <td>0.206084</td>\n",
       "      <td>0.649957</td>\n",
       "      <td>1.093829</td>\n",
       "      <td>-0.126821</td>\n",
       "      <td>-0.459725</td>\n",
       "      <td>-1.569407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>0.940315</td>\n",
       "      <td>0.747687</td>\n",
       "      <td>-0.041919</td>\n",
       "      <td>0.731962</td>\n",
       "      <td>-1.017976</td>\n",
       "      <td>0.344460</td>\n",
       "      <td>-0.213207</td>\n",
       "      <td>2.093275</td>\n",
       "      <td>-0.838265</td>\n",
       "      <td>0.031650</td>\n",
       "      <td>0.197883</td>\n",
       "      <td>-1.050549</td>\n",
       "      <td>-0.030687</td>\n",
       "      <td>-1.894630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RP5-857K21.11</td>\n",
       "      <td>2.203644</td>\n",
       "      <td>0.739910</td>\n",
       "      <td>-0.498635</td>\n",
       "      <td>0.852505</td>\n",
       "      <td>0.514720</td>\n",
       "      <td>0.176935</td>\n",
       "      <td>-1.061610</td>\n",
       "      <td>0.852505</td>\n",
       "      <td>-1.286800</td>\n",
       "      <td>-1.061610</td>\n",
       "      <td>-0.611230</td>\n",
       "      <td>0.402125</td>\n",
       "      <td>-0.160850</td>\n",
       "      <td>-1.061610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15852</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>0.615087</td>\n",
       "      <td>1.549672</td>\n",
       "      <td>-1.084553</td>\n",
       "      <td>1.480122</td>\n",
       "      <td>-0.386875</td>\n",
       "      <td>0.015214</td>\n",
       "      <td>-0.504241</td>\n",
       "      <td>0.791137</td>\n",
       "      <td>-0.378181</td>\n",
       "      <td>-0.919371</td>\n",
       "      <td>1.197573</td>\n",
       "      <td>-0.806351</td>\n",
       "      <td>0.086938</td>\n",
       "      <td>-1.656171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15853</th>\n",
       "      <td>MT-TE</td>\n",
       "      <td>2.744216</td>\n",
       "      <td>0.154512</td>\n",
       "      <td>-0.424363</td>\n",
       "      <td>-0.302495</td>\n",
       "      <td>1.555999</td>\n",
       "      <td>-0.759501</td>\n",
       "      <td>-0.698567</td>\n",
       "      <td>0.154512</td>\n",
       "      <td>-0.759501</td>\n",
       "      <td>-0.393896</td>\n",
       "      <td>-0.028291</td>\n",
       "      <td>-0.546232</td>\n",
       "      <td>-0.789969</td>\n",
       "      <td>0.093578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15854</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>0.632780</td>\n",
       "      <td>1.498360</td>\n",
       "      <td>-0.696590</td>\n",
       "      <td>1.327776</td>\n",
       "      <td>-0.845693</td>\n",
       "      <td>0.188345</td>\n",
       "      <td>-0.329790</td>\n",
       "      <td>1.552492</td>\n",
       "      <td>-0.159524</td>\n",
       "      <td>-0.640012</td>\n",
       "      <td>0.383497</td>\n",
       "      <td>-1.075408</td>\n",
       "      <td>-0.077848</td>\n",
       "      <td>-1.758385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15855</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>1.625994</td>\n",
       "      <td>1.278074</td>\n",
       "      <td>-0.063904</td>\n",
       "      <td>1.153817</td>\n",
       "      <td>0.420699</td>\n",
       "      <td>-1.033110</td>\n",
       "      <td>-1.120090</td>\n",
       "      <td>1.116540</td>\n",
       "      <td>-1.008258</td>\n",
       "      <td>-0.150884</td>\n",
       "      <td>0.383422</td>\n",
       "      <td>-0.511230</td>\n",
       "      <td>-1.107664</td>\n",
       "      <td>-0.983407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15856</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>2.108175</td>\n",
       "      <td>0.590943</td>\n",
       "      <td>-0.725899</td>\n",
       "      <td>0.734078</td>\n",
       "      <td>0.963095</td>\n",
       "      <td>-0.554137</td>\n",
       "      <td>-0.296494</td>\n",
       "      <td>-0.267867</td>\n",
       "      <td>-1.155305</td>\n",
       "      <td>-1.212559</td>\n",
       "      <td>1.077603</td>\n",
       "      <td>-1.098051</td>\n",
       "      <td>-0.611391</td>\n",
       "      <td>0.447808</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15309 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name     HC1_1     HC2_1     HC3_1   B.I.1_1  B.II.2_1  \\\n",
       "1           MTND2P28 -0.082342  0.017687 -0.864533  0.063601 -0.841576   \n",
       "2       hsa-mir-6723  1.481961  1.093755 -0.717873  1.913301  0.144807   \n",
       "3       RP5-857K21.7  0.649957  0.095116  0.428020  0.095116 -1.902312   \n",
       "4           MTATP6P1  0.940315  0.747687 -0.041919  0.731962 -1.017976   \n",
       "5      RP5-857K21.11  2.203644  0.739910 -0.498635  0.852505  0.514720   \n",
       "...              ...       ...       ...       ...       ...       ...   \n",
       "15852         MT-ND6  0.615087  1.549672 -1.084553  1.480122 -0.386875   \n",
       "15853          MT-TE  2.744216  0.154512 -0.424363 -0.302495  1.555999   \n",
       "15854         MT-CYB  0.632780  1.498360 -0.696590  1.327776 -0.845693   \n",
       "15855          MT-TT  1.625994  1.278074 -0.063904  1.153817  0.420699   \n",
       "15856          MT-TP  2.108175  0.590943 -0.725899  0.734078  0.963095   \n",
       "\n",
       "       C.II.1_1  C.III.1_1     HC1_2     HC2_2     HC3_2   B.I.1_2  B.II.2_2  \\\n",
       "1      2.116648   1.324618  0.024246 -0.646438 -0.712031 -0.344712 -1.062951   \n",
       "2     -0.588471  -0.933543  0.748683 -0.415935 -1.235481  0.533013 -0.415935   \n",
       "3      0.206084  -1.125534  1.759638  0.206084  0.649957  1.093829 -0.126821   \n",
       "4      0.344460  -0.213207  2.093275 -0.838265  0.031650  0.197883 -1.050549   \n",
       "5      0.176935  -1.061610  0.852505 -1.286800 -1.061610 -0.611230  0.402125   \n",
       "...         ...        ...       ...       ...       ...       ...       ...   \n",
       "15852  0.015214  -0.504241  0.791137 -0.378181 -0.919371  1.197573 -0.806351   \n",
       "15853 -0.759501  -0.698567  0.154512 -0.759501 -0.393896 -0.028291 -0.546232   \n",
       "15854  0.188345  -0.329790  1.552492 -0.159524 -0.640012  0.383497 -1.075408   \n",
       "15855 -1.033110  -1.120090  1.116540 -1.008258 -0.150884  0.383422 -0.511230   \n",
       "15856 -0.554137  -0.296494 -0.267867 -1.155305 -1.212559  1.077603 -1.098051   \n",
       "\n",
       "       C.II.1_2  C.III.1_2  \n",
       "1      1.647661  -0.639879  \n",
       "2     -0.502203  -1.106079  \n",
       "3     -0.459725  -1.569407  \n",
       "4     -0.030687  -1.894630  \n",
       "5     -0.160850  -1.061610  \n",
       "...         ...        ...  \n",
       "15852  0.086938  -1.656171  \n",
       "15853 -0.789969   0.093578  \n",
       "15854 -0.077848  -1.758385  \n",
       "15855 -1.107664  -0.983407  \n",
       "15856 -0.611391   0.447808  \n",
       "\n",
       "[15309 rows x 15 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# zscore across samples per gene with ddof=1 default for sp.stats.zscore\n",
    "zscored_df_1 = kuehn_df.loc[:, 'HC1_1':].apply(sp.stats.zscore, axis=1, ddof=1)\n",
    "zscored_df_1.insert(0, 'gene_name', kuehn_df['gene_name'])\n",
    "zscored_df_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zscored across samples per gene\n",
    "# sns.set_context('paper', font_scale = 0.5) \n",
    "# sns.clustermap(zscored_df_1.loc[:, 'HC1_1':], \n",
    "#                xticklabels=zscored_df_1.columns[1:], \n",
    "#                yticklabels=zscored_df_1['gene_name'],\n",
    "#                cmap='coolwarm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DAL",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
